---
title: "Differential methylation analysis"
author: "Andrew Y.F. Li Yim"
date: "February 25, 2020"
output: html_document
editor_options: 
  chunk_output_type: console
---

# Introduction

In this report I will perform differential methylation analyses on DNA methylation data (Illumina HumanMethylation 450k BeadChip array) obtained from CD14+ monocytes, which were isolated from CD patients (N = 16) and non-CD volunteers (N = 8). The CD patients were further subdivided into CD-active (N = 8) and remissive CD (N = 8), where activity was defined based on global patient assessment.

In preparation of the actual analysis, I will need to prepare the necessary objects. First, the necessary packages need to be imported.

```{r packages}
require(Biobase)
require(org.Hs.eg.db)
require(pheatmap)
require(minfi)
require(Cairo)
require(ggplot2)
require(ggrepel)
require(dplyr)
require(plotly)
require(NDlib) # Can be found on my Github
```

```{r setup}
samplesDir <- file.path("data", "samples")
commondataDir <- file.path("/home", "ayliyim", "archive", "common_data")

isodate <- "200225"
#isodate <- strftime(Sys.time(), "%y%m%d")

outputDir <- file.path("output", "01_DM_450k_analyses", isodate)
dir.create(outputDir)
manuscriptDir <- file.path("docs", "manuscript", "data", isodate)
dir.create(manuscriptDir)
```

## Samples

Samples will be imported using [`minfi`](https://bioconductor.org/packages/release/bioc/html/minfi.html).

```{r samples import}
samplesheet <- read.metharray.sheet(base = samplesDir, pattern = "samplesheet_450k_PRJ0000002_CDMON_V6")
samplesheet$Basename <- samplesheet$Path
samplesheet$Sex <- "F"
```

```{r samples per donor}
table(samplesheet$Donor_ID)
```

There are two samples from donor `CDACT-1`. From the information provided by my colleagues, it appeared that they *might* have messed up one sample, but it was not entirely sure what had occurred. So both samples were run from the same individual on the methylation array. 

I first import the data and preprocess it using the `preprocessFunnorm` normalization method, which performs background correction and quantile normalization. An unfortunate side-effect of `preprocessFunnorm` is that the resulting M-values have `-Inf` values, which will need to be removed. I also perform some initial filtering of the data by removing probes annotated with known genetic variants and heterogeneous probes.

```{r data import and preprocessing}
rgset <- read.metharray.exp(targets = samplesheet)
gmset <- preprocessFunnorm(rgSet = rgset)
gmset_nosnpcpg <- gmset[with(getSnpInfo(gmset), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
gmset_nosnpsbe <- gmset_nosnpcpg[with(getSnpInfo(gmset_nosnpcpg), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
gmset_nosex <- gmset_nosnpsbe[-which(getAnnotation(gmset_nosnpsbe)$chr %in% c("chrX", "chrY")),]

pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChip450k", "Non-specific-probes-Illumina450k.csv"), stringsAsFactors = F, header = T)[,1]
gmset_filtered <- gmset_nosex[!names(gmset_nosex) %in% pmprobes,]

mvals <- getM(gmset_filtered)

gmset_filtered <- gmset_filtered[-which(abs(mvals) == Inf, arr.ind = T)[,1],]
pData(gmset_filtered)$Severity <- factor(pData(gmset_filtered)$Severity, levels = c("Non_CD", "Remissive", "Active"))

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

## Annotations

To annotate the observed CpGs, I use a variety of sources, one of which is the promoter capture Hi-C data obtained from [Javierre et al. 2016](https://www.ncbi.nlm.nih.gov/pubmed/27863249). Using this data, I can investigate whether differentially methylated loci occur in enhancer regions and perhaps more importantly, which gene it might enhance. 

```{r promoter capture hic}
pchic <- read.csv(file.path(commondataDir, "enhancers/javierre2016/ActivePromoterEnhancerLinks_hg19.tsv"), sep = "\t")
pchic_gr <- makeGRangesFromDataFrame(pchic, keep.extra.columns = T, seqnames.field = "oeChr", start.field = "oeSt", end.field = "oeEnd")
```

## Visualization

The following functions are mostly used for visualization purposes only.

```{r bespoke functions}
source(file.path("src", "dmr_plot.R"))
source(file.path("src", "dmg_plot.R"))
source(file.path("src", "dmcomp_plot.R"))
```

# Quality control and exploratory data

The first step in the analysis represents a sanity check of the data. Based on the given metadata I should see the following:

* There are two samples from donor `CDACT1`.
* All samples were derived from female donors.
* Samples pertain monocytes.

For quality control, I utilized [`MethylAid`](https://bioconductor.org/packages/release/bioc/html/MethylAid.html), [`shinyMethyl`](https://bioconductor.org/packages/release/bioc/html/shinyMethyl.html), [`ewastools`](https://github.com/hhhh5/ewastools) and `FlowSorted.Blood.450k` in combination with some built-in functions in `minfi`. Both `MethylAid` and `shinyMethyl` provide similar functionality, but certain aspects work better in one or the other. For example, I like the thresholds provided from `MethylAid` and the more up-to-date feel it provides with newer EPIC arrays, but I like the interface of `shinyMethyl` more as well as its more expanded functionality. In addition, I use `ewastools` to identify whether the samples from the same donor (`CDACT-1`) are actually derived from the same donor. In short, `ewastools` uses probes that bind known genetic variants to build a profile per patient. Conversely, I can also confirm that samples derived from different samples were indeed not derived (or mixed). Finally, `FlowSorted.Blood.450k` are basically 450k arrays run on flow sorted blood cells to investigate the blood cell distribution.

Note that the next section is currently commented out simply because running it accidentally takes a lot of time and memory.

```{r eda qc setup}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)
```

```{r qc packages}
require(MethylAid)
require(shinyMethyl)
require(ewastools)
require(FlowSorted.Blood.450k)
```

```{r methylaid}
# methylaid_summarized <- summarize(samplesheet)
# visualize(methylaid_summarized)
```

According to `MethylAid`, all samples pass the pre-imposed quality control thresholds, which is reassuring. 

```{r shinymethyl}
# require(shinyMethyl)
# shinymethyl_summarized <- shinySummarize(rgset)
# runShinyMethyl(shinymethyl_summarized)
```

We find from `shinyMethyl` that all samples are female, which is in accordance with the provided information. We can better visualize this as following:

```{r Annotated and predicted sex}
require(ggplot2)
require(dplyr)
require(ggrepel)

pData(gmset_filtered) %>% 
  data.frame %>% 
  select(Sample_ID, Sex, predictedSex, xMed, yMed) %>%
  mutate(Sample_label = ifelse(Sex == predictedSex, "", Sample_ID)) %>% 
  ggplot(aes(x = xMed, y = yMed, col = Sex, label = Sample_label)) +
  geom_text_repel(show.legend = F) +
  geom_point(size = 4) +
  theme_bw() +
  coord_fixed() +
  ggtitle("Sex chromosome probe median intensities") +
  xlab("chrX") +
  ylab("chrY") +
  theme(text = element_text(size=20))
```

```{r ewastools}
# ewastools_samples <- read_idats(samplesheet$Basename)
# ewastools_betas <- dont_normalize(ewastools_samples)
# ewastools_snps <- call_genotypes(ewastools_betas[ewastools_samples$manifest[probe_type == 'rs', index], ], learn = F)
# check_snp_agreement(genotypes = ewastools_snps$gamma, weights = 1 - ewastools_snps$outliers, donor_ids = samplesheet$Donor_ID, sample_ids = samplesheet$Sample_ID)
```

According to `ewastools`, we observe that the samples belonging to `CDACT-1` were indeed obtained from the same donor, confirming the information from my colleague.

```{r bloodcell distribution estimation}
cellCounts <- estimateCellCounts(rgSet = rgset, 
                                 compositeCellType = "Blood",
                                 processMethod = "auto", 
                                 probeSelect = "auto")
cellCounts <- data.frame(cellCounts, 
                         severity = pData(rgset)$Severity, 
                         array = rownames(cellCounts))

cellCounts_melt <- reshape2::melt(cellCounts, id.vars = c("severity", "array"), id.vals = c("Response"))

cellCounts_plot <- ggplot(cellCounts_melt, aes(x = severity, y = value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter() +
  facet_wrap(~variable) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 2000, height = 2000, type = "pdf", file = file.path(edaDir, "estimatedCellCounts.pdf"), bg = "white", units = "px", dpi = 120)
print(cellCounts_plot)
dev.off()
```

In general, the cell distribution estimation tool accurately guesses that the samples consist of monocytes. While there appears to be some variation, there does not appear to be a strong association with phenotype except for the granulocytes, where the CD patients appear to display an increased presence. In the CD8T, CD4T and monocytes, there appears to be a sample however that appears different. Naturally, the question is whether that sample is the same for all.

```{r outlier sample}
CD4T_unique_sxs_ID <- cellCounts_melt %>% 
  filter(variable == "CD4T" & value > 0.25) %>% 
  select(array)

CD8T_unique_sxs_ID <- cellCounts_melt %>% 
  filter(variable == "CD8T" & value > 0.125) %>% 
  select(array)

Mono_unique_sxs_ID <- cellCounts_melt %>% 
  filter(variable == "Mono" & value < 0.25) %>% 
  select(array)

unique(CD4T_unique_sxs_ID$array, CD8T_unique_sxs_ID$array, Mono_unique_sxs_ID$array)
```

We indeed observe that it is the same sample, namely `3998609037_R01C01`. Note that we could infer that the outlier sample in every cell type should have been the same based on the fact that the prediction responses together approximately add up to 1.

```{r sample ID}
pData(rgset)[CD4T_array_ID$array, "Sample_ID"]
```

We have established that sample `HC1` behaves distinct from the rest **based on the estimated cell distribution**. The next  question is therefore whether this is only for the bloodcell-associated CpGs, or whether this is something genome-wide. To this end, we can perform principal component analysis (PCA).

```{r pca}
mvals_dm <- mvals - rowMeans(mvals)
mvals_svd <- svd(t(mvals_dm))

mvals_svd_df <- data.frame(Sample_ID = pData(gmset_filtered)$Sample_ID,
                           Origin = pData(gmset_filtered)$Donor_ID,
                           Age = pData(gmset_filtered)$Age,
                           Severity = pData(gmset_filtered)$Severity,
                           PC1 = mvals_svd$u[,1],
                           PC2 = mvals_svd$u[,2])

mvals_svd_plotobj <- ggplot(mvals_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Severity = Severity, Age = Age)) +
  geom_point(shape = 21, size = 4, aes(fill = Severity)) +
  theme_bw() +
  labs(title = "Principal component analysis") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

ggplotly(mvals_svd_plotobj)
```

We indeed see that sample `HC1` behaves distinct from the rest. Note that `HC6` also displays some degree of "uniqueness". I will keep `HC6` in for now, but remove `HC1`.

```{r removal HC1}
gmset_filtered <- gmset_filtered[, pData(gmset_filtered)$Sample_ID != "HC1"]

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

In addition, we will aggregate the data obtained from the two `CDACT1` samples.

```{r aggregation CDACT1}
mvals_filtered <- data.frame(CDACT1 = rowMeans(mvals[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), mvals[, 3:ncol(mvals)])
betas_filtered <- data.frame(CDACT1 = rowMeans(betas[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), betas[, 3:ncol(betas)])

pdata_filtered <- pData(gmset_filtered)[-1,]
colnames(mvals_filtered) <- colnames(betas_filtered) <- rownames(pdata_filtered)
```

If we generate a new PCA, we should not be able to see any (severe) outlier.

```{r pca filtered}
mvals_filtered_dm <- mvals_filtered - rowMeans(mvals_filtered)
mvals_filtered_svd <- svd(t(mvals_filtered_dm))

mvals_filtered_svd_df <- data.frame(Sample_ID = pdata_filtered$Sample_ID,
                                    Origin = pdata_filtered$Donor_ID,
                                    Age = pdata_filtered$Age,
                                    Severity = pdata_filtered$Severity,
                                    PC1 = mvals_filtered_svd$u[,1],
                                    PC2 = mvals_filtered_svd$u[,2])

mvals_filtered_svd_plotobj <- ggplot(mvals_filtered_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Severity = Severity, Age = Age)) +
  geom_point(shape = 21, size = 4, aes(fill = Severity)) +
  theme_bw() +
  labs(title = "Principal component analysis") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

ggplotly(mvals_filtered_svd_plotobj)
```

# Differential methylation

Prepare annotation for the eventual interpretation.

```{r annotation preparation}
gmset_anno <- getAnnotation(gmset_filtered)
gmset_anno_gr <- makeGRangesFromDataFrame(gmset_anno, keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
```

We store the DMPs, DMRs and DMGs in separate directories.

```{r dm output directories}
dmpDir <- file.path(outputDir, "dmps")
dir.create(dmpDir)
dmrDir <- file.path(outputDir, "dmrs")
dir.create(dmrDir)
dmgDir <- file.path(outputDir, "dmgs")
dir.create(dmgDir)
```

```{r setup differential analyses}
require(limma)
```

## CD vs non-CD

First we compare CD with non-CD.First we compare the global differences between CD and non-CD where we correct for age. We do not need to correct for sex as the samples were obtained from an all-female cohort.

```{r dm analysis preparation cvn}
pdata_filtered$Status1 <- factor(pdata_filtered$Status1, levels = c("Non_CD", "CD"))

design_cvn <- model.matrix(~Status1 + Age, data = pdata_filtered)

mvals_lmfit_cvn <- lmFit(mvals_filtered, design = design_cvn)
betas_lmfit_cvn <- lmFit(betas_filtered, design = design_cvn)

mvals_ebayes <- eBayes(mvals_lmfit_cvn)
betas_ebayes <- eBayes(betas_lmfit_cvn)
```

### DMPs

We first investigate whether we can identify particular probes that bind CpG loci that are differentially methylated, which we call differentially methylated probes (DMPs). 

```{r dm analysis cvn}
CDvNon_CD <- topTable(mvals_ebayes, coef = "Status1CD", number = "Inf", adjust.method = "BH")
CDvNon_CD <- cbind(CDvNon_CD, Beta = coefficients(betas_ebayes)[rownames(CDvNon_CD), "Status1CD"], gmset_anno[rownames(CDvNon_CD), ])

write.csv(CDvNon_CD, file.path(dmpDir, "CDvNon_CD.csv"))

for(i in 1:10){
  Cairo(width = 600, height = 800, type = "pdf", file = file.path(dmpDir, paste0(i, "_", rownames(CDvNon_CD)[i], "_CDvNon_CD.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(CDvNon_CD)[i], betas = betas_filtered, factor_interest = pdata_filtered$Status1, enlarged = F, type = "SE", legend = F) + 
          theme(panel.grid.major=element_blank(),
                panel.grid.minor=element_blank()))
  dev.off()
}
```

Note that we find no statistically significant differences between CD and non-CD individuals. However, when we visualize the difference between the CD and non-CD individuals, there appear to be observable differences with a clear mean difference. 

```{r heatmap cvn}
hm_anno_t50_cvn <- data.frame(Status = pdata_filtered$Status1, row.names = rownames(pdata_filtered))

betas_t50_cvn <- betas_filtered[rownames(CDvNon_CD[1:50,]),]
betas_t50_cvn_dm <- betas_t50_cvn-rowMeans(betas_t50_cvn)

pheatmap(betas_t50_cvn_dm, annotation_col = hm_anno_t50_cvn)
```

From the heatmap, it is clear that among the top 50 DMPs there are noticable differences. The lack of statistical significance does not necessarily make it biologically insignificant. 

We know that the DNA methylation status of neighboring CpGs are typically correlated, we would therefore want to know whether the most significant DMP is correlated. 

```{r methylation surrounding most dmps cvn}
#cg17206772
cg17206772_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[1])
cg17206772_area_gr <- gmset_anno_gr[(cg17206772_area-5):(cg17206772_area+5),]
CDvNon_CD[names(cg17206772_area_gr),]

dmr_plot(dmr_gr = range(cg17206772_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)

#cg16921727
cg16921727_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[2])
cg16921727_area_gr <- gmset_anno_gr[(cg16921727_area-5):(cg16921727_area+5),]
CDvNon_CD[names(cg16921727_area_gr),]

dmr_plot(dmr_gr = range(cg16921727_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)
```

For `cg17206772` we observe differential methylation but do not observe such differences among its neighboring CpGs, whose nearest neighbors are approximately 1000 up and downstream. Similarly, for `cg16921727` we observe no visible differential methylation among its neighbors. 

### DMGs

While it is clear that there is no statistically significant effect given the sample size at hand, we do observe some visible differences in methylation. However, individual probes by themselves are often not very informative. It might be more useful to identify entities that are differentially methylated. To this end, I sought to identify genes that are enriched for differential methylation. Harnessing the annotation from Illumina, we can perform some meta-analytical approach to identify genes with multiple probes whose nominal p-values are low. A similar approach was recently described for finding genes of differential expression by [Yi et al. 2018](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1419-z). We have to implement some modifications however, the Fisher or Lancaster method assumes independence between the elements of an entity. While such a case can somewhat be made for individual transcripts, we cannot make the same assumption for DNA methylation as mentioned before. To that end, we utilize Brown's function instead as implemented in [`EmpiricalBrownsMethod`](https://bioconductor.org/packages/release/bioc/html/EmpiricalBrownsMethod.html), which accommodates for the correlation structure. Note that the implemented tool simply looks for genes enriched for CpGs with low p-values and does not necessarily find consecutive CpGs with low p-values, that is something we will explore next.

```{r differentially methylated genes cvn}
CDvNon_CD_genes <- lapply(strsplit(CDvNon_CD$UCSC_RefGene_Name, ";"), unique)
names(CDvNon_CD_genes) <- rownames(CDvNon_CD)
CDvNon_CD_genes_nz <- CDvNon_CD_genes[which(lapply(CDvNon_CD_genes, length) != 0)]

CDvNon_CD_genes_nz_df <- data.frame(CpG = gsub("(cg[0-9]{8})[0-9]", "\\1", names(unlist(CDvNon_CD_genes_nz))), 
                                    Gene = unlist(CDvNon_CD_genes_nz))
CDvNon_CD_genes_nz_df$pvalue <- CDvNon_CD[CDvNon_CD_genes_nz_df$CpG, "P.Value"]

CDvNon_CD_genes_nz_list <- split(CDvNon_CD_genes_nz_df, CDvNon_CD_genes_nz_df$Gene)

CDvNon_CD_fisher <- lapply(CDvNon_CD_genes_nz_list, function(gene){
  #aggregation::fisher(gene$pvalue)
  unlist(EmpiricalBrownsMethod::empiricalBrownsMethod(data_matrix = betas_filtered[gene$CpG,],
                                                      p_values = gene$pvalue,
                                                      extra_info = T))
})

rm(CDvNon_CD_genes_nz_list)

CDvNon_CD_ma <- data.frame(Gene = names(CDvNon_CD_fisher), do.call(rbind, CDvNon_CD_fisher))
colnames(CDvNon_CD_ma) <- c("Gene", "Brown_pval", "Fisher_pval", "Scale_Factor_C", "DF")
CDvNon_CD_ma$Brown_padj <- p.adjust(CDvNon_CD_ma$Brown_pval)
CDvNon_CD_ma <- CDvNon_CD_ma[order(CDvNon_CD_ma$Brown_pval),]
CDvNon_CD_ma_sig <- CDvNon_CD_ma[which(CDvNon_CD_ma$Brown_padj<0.05),]

write.csv(CDvNon_CD_ma, file.path(dmgDir, "CDvNon_CD.csv"))
```

Naturally, the DMGs might make more sense if they are visualized.

```{r plotting dmgs cvn}
CDvNon_CD_gr <- makeGRangesFromDataFrame(CDvNon_CD, keep.extra.columns = T, start.field = "pos", end.field = "pos", seqnames.field = "chr")

# for(i in 10:length(CDvNon_CD_ma_sig$Gene)){
#   Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(i, "_", CDvNon_CD_ma_sig$Gene[i], "_CDvnonCD.pdf")), bg = "white", units = "px", dpi = 120)
#   print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[i], tophits_gr = CDvNon_CD_gr, smooth = T, significance = F))
#   dev.off()
# }

# Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(3, "_", CDvNon_CD_ma_sig$Gene[3], "_CDvnonCD.pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[3], tophits_gr = CDvNon_CD_gr[start(CDvNon_CD_gr) > 24383000], smooth = T, significance = F))
# dev.off()

# Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(10, "_", CDvNon_CD_ma_sig$Gene[10], "_CDvnonCD.pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[10], tophits_gr = CDvNon_CD_gr[start(CDvNon_CD_gr) < 107310000], smooth = T, significance = F))
# dev.off()
```

Note that for some of the DMGs we notice the grey area (standard error) to exceed the possible range [-1,1]. This is something I have not found a proper solution for.

For most of the genes, we find that the difference in methylation occurs consecutively, with many of the consecutive regions occurring within the vicinity of the transcription start site (TSS). Under the observation that DNA methylation is frequently inversely correlated with gene expression, that could suggest differential expression. However, different types of experiments are necessary for this. 

#### Further annotations

We next investigated whether the difference in methylation occurred in any of the enhancers regions annotated by Javierre et al.

```{r dmg cvn enhancers}
dmg_CDvNon_CD_enhancers <-lapply(CDvNon_CD_ma_sig$Gene, function(dmg){
  gene_gr <- CDvNon_CD_gr[grep(dmg, CDvNon_CD_gr$UCSC_RefGene_Name),]
  findOverlaps(pchic_gr, gene_gr)
})

CDvNon_CD_ma_sig[11,]
```

We find only that one DMG to overlap an enhancer region, namely SPI1. Naturally, we would want to know in which cell types this enhancer is active.

```{r spi1 enhancer gene}
pchic_gr[unique(queryHits(dmg_CDvNon_CD_enhancers[[11]]))]
```

SPI1 appears to display enhancer functions, but according to Ensembl Regulatory in erythroblasts, neutrophils and M0 and M1 macrophages only.

#### Gene set analyses

We can further identify the biological implications of the differences in methylation by performing overrepresentation analyses. For this, we use [`DOSE`](https://bioconductor.org/packages/release/bioc/html/DOSE.html) and [`clusterProfiler`](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html), where we wil perform gene set overrepresentation against the KEGG database.

```{r gene set analyses setup}
require(DOSE)
require(clusterProfiler)
```

While we can annotate most genes automatically, some of the gene symbols cannot easily be annotated and must hence be done manually.

```{r gene set dmgs cvn}
CDvNon_CD_ma_sig$Entrez <- mapIds(Homo.sapiens, 
                                  keys = as.character(CDvNon_CD_ma_sig$Gene),
                                  column = "ENTREZID",
                                  keytype = "SYMBOL",
                                  multiVals = "first")
CDvNon_CD_ma_sig["C6orf25", "Entrez"] <- 80739
CDvNon_CD_ma_sig["C22orf26", "Entrez"] <- 55267
CDvNon_CD_ma_sig["PNMAL2", "Entrez"] <- 57469
CDvNon_CD_ma_sig["C10orf26", "Entrez"] <- 54838

CDvNon_CD_ma_sig_kegg <- enrichKEGG(gene = CDvNon_CD_ma_sig$Entrez,
                                    organism = 'hsa')
```

We unfortunately do not find any significantly enriched gene set. Idea for another time: we might be losing much of the signal simply by looking at statistically significant genes only. It might make more sense to take the mean the methylation per promoter and use that as effect size in gene set enrichment analyses to see whether particular gene sets are hyper or hypomethylated.

### DMRs

As alluded to previously, our DMG analysis did not necessarily yield consecutive regions of differential methylation. To do so, I would use [`DMRcate`](https://bioconductor.org/packages/release/bioc/html/DMRcate.html). The nice thing about `DMRcate` is that I can essentially feed it the same model matrix (and contrasts) as used in the DMP analysis.

```{r dmr analysis setup}
dmrcvnDir <- file.path(dmrDir, "cvn")
dir.create(dmrcvnDir)

require(DMRcate)
```

```{r dmr analysis cvn}
dmr_anno_CDvNon_CD <- cpg.annotate(object = as.matrix(mvals_filtered),
                                   datatype = "array",
                                   arraytype = "450K",
                                   what = "M",
                                   analysis.type = "differential",
                                   coef = "Status1CD",
                                   design = design_cvn)
dmr_CDvNon_CD <- dmrcate(dmr_anno_CDvNon_CD, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_CDvNon_CD_gr <- extractRanges(dmr_CDvNon_CD, genome = "hg19")

write.csv(as.data.frame(dmr_CDvNon_CD_gr), file.path(dmrDir, "dmrs_CDvNon_CD.csv"))
```

Once again, no statistically significant DMR was observed. However, we would still want to visualize the most significant DMR to see whether it was nonetheless observable. 

```{r dmr cvn visualization}
dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], 
         meth_data = betas_filtered, 
         anno_gr = gmset_anno_gr, 
         meth_groups = pdata_filtered$Status1, 
         flanks = 0)
```

Visualization does appear to suggest that the CD group is different from the non-CD group. It might be worthwhile to investigate this in future confirmatory studies. 

## ACT vs REM

Within the CD patients, we would want to compare the samples from CD-active with CD-remissive. 

```{r dm analysis preparation avr}
dmpavrDir <- file.path(dmpDir, "avr")
dir.create(dmpavrDir)

pdata_avr <- pdata_filtered[pdata_filtered$Status1 == "CD", ]
pdata_avr$Severity <- factor(pdata_avr$Severity, c("Remissive", "Active"))
mvals_avr <- mvals_filtered[,rownames(pdata_avr)]
betas_avr <- betas_filtered[,rownames(pdata_avr)]
```

Medication is naturally a very important aspect of CD and is likely a very strong reason underlying activity. It would therefore be prudent to investigate how medication might relate to the observations made. First question is, can we control for it in our model? 

```{r avr medication}
table(pdata_avr[,c("Severity", "Medication")])
```

Notably, most patients are on some kind of medication. However, 2 of the CD-active patients are not on medication. This is rather unfortunate as adding it to the model would require an additional covariate thereby decreasing the degrees of freedom. Instead of controlling for it, I will check post-hoc whether the CD-active patients on no medication behave distinct.

### DMPs

```{r dm analysis avr}
design_avr <- model.matrix(~Severity + Age, data = pdata_avr)

mvals_lmfit_avr <- lmFit(mvals_avr, design = design_avr)
betas_lmfit_avr <- lmFit(betas_avr, design = design_avr)

mvals_ebayes_avr <- eBayes(mvals_lmfit_avr)
betas_ebayes_avr <- eBayes(betas_lmfit_avr)

ACTvREM <- topTable(mvals_ebayes_avr, coef = "SeverityActive", number = "Inf", adjust.method = "BH")
ACTvREM <- cbind(ACTvREM, Beta = coefficients(betas_ebayes_avr)[rownames(ACTvREM), "SeverityActive"], gmset_anno[rownames(ACTvREM), ])

write.csv(ACTvREM, file.path(dmpavrDir, "ACTvREM.csv"))

for(i in 1:10){
  Cairo(width = 600, height = 800, type = "pdf", file = file.path(dmpavrDir, paste0(i, "_", rownames(ACTvREM)[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(ACTvREM)[i], betas = betas_avr, factor_interest = pdata_avr$Severity, enlarged = F, type = "SE", legend = F) + 
          theme(panel.grid.major=element_blank(),
                panel.grid.minor=element_blank()))
  dev.off()
}
```

```{r heatmap avr}
hm_anno_t50_avr <- data.frame(Status = pdata_avr$Severity, row.names = rownames(pdata_avr))
hm_col_anno_t50_avr <- data.frame(pdata_avr$Sample_ID, row.names = rownames(pdata_avr))

betas_t50_avr <- betas_avr[rownames(ACTvREM[1:50,]),]
betas_t50_avr_dm <- betas_t50_avr-rowMeans(betas_t50_avr)

Cairo(width = 2000, height = 1500, type = "pdf", file = file.path(dmpavrDir, "heatmap_ACTvREM_top50.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap(betas_t50_avr_dm, annotation_col = hm_anno_t50_avr, labels_col = hm_col_anno_t50_avr$pdata_avr.Sample_ID)
dev.off()
```

Similar to CD vs non-CD, we find no statistically significant differences between CD-active and CD-remissive. Again, when we visualize the difference between the CD and non-CD individuals, there appear to be observable differences, which is perhaps more visible from the heatmap. 

### DMGs

Similar to what we have done before, we will try to find genes that are enriched for DMPs.

```{r differentially methylated genes avr}
dmgavrDir <- file.path(dmgDir, "avr")
dir.create(dmgavrDir)

ACTvREM_genes <- lapply(strsplit(ACTvREM$UCSC_RefGene_Name, ";"), unique)
names(ACTvREM_genes) <- rownames(ACTvREM)
ACTvREM_genes_nz <- ACTvREM_genes[which(lapply(ACTvREM_genes, length) != 0)]

ACTvREM_genes_nz_df <- data.frame(CpG = gsub("(cg[0-9]{8})[0-9]", "\\1", names(unlist(ACTvREM_genes_nz))), 
                                  Gene = unlist(ACTvREM_genes_nz))
ACTvREM_genes_nz_df$pvalue <- ACTvREM[ACTvREM_genes_nz_df$CpG, "P.Value"]

ACTvREM_genes_nz_list <- split(ACTvREM_genes_nz_df, ACTvREM_genes_nz_df$Gene)

ACTvREM_fisher <- lapply(ACTvREM_genes_nz_list, function(gene){
  #aggregation::fisher(gene$pvalue)
  unlist(EmpiricalBrownsMethod::empiricalBrownsMethod(data_matrix = betas_filtered[gene$CpG,],
                                                      p_values = gene$pvalue,
                                                      extra_info = T))
})

rm(ACTvREM_genes_nz_list)

ACTvREM_ma <- data.frame(Gene = names(ACTvREM_fisher), do.call(rbind, ACTvREM_fisher))
colnames(ACTvREM_ma) <- c("Gene", "Brown_pval", "Fisher_pval", "Scale_Factor_C", "DF")
ACTvREM_ma$Brown_padj <- p.adjust(ACTvREM_ma$Brown_pval)
ACTvREM_ma <- ACTvREM_ma[order(ACTvREM_ma$Brown_pval),]
ACTvREM_ma_sig <- ACTvREM_ma[which(ACTvREM_ma$Brown_padj<0.05),]

write.csv(ACTvREM_ma, file.path(dmgavrDir, "ACTvREM.csv"))
```

Now that we have our genes of interest, the question is naturally whether medication affects the methylation signal of the DMGs. We can perform PCA on only these CpGs and investigate whether the CD-active patients on no medication cluster separately from the other CD-active, or even all CD patients.

```{r medication effect dmgs avr}
ACTvREM_ma_sig_cpgs <- unlist(lapply(ACTvREM_ma_sig$Gene, function(gene){
  ACTvREM$Name[grep(gene, ACTvREM$UCSC_RefGene_Name)]
}))

mvals_avr_dmeg <- mvals_avr[ACTvREM_ma_sig_cpgs,]
mvals_avr_dmeg <- mvals_avr_dmeg-rowMeans(mvals_avr_dmeg)

mvals_avr_dmeg_svd <- svd(t(mvals_avr_dmeg))
mvals_avr_dmeg_pct_var <- mvals_avr_dmeg_svd$d/sum(mvals_avr_dmeg_svd$d)*100

mvals_avr_dmeg_svd_df <- data.frame(Sample_ID = pdata_avr$Sample_ID,
                                    Medication = pdata_avr$Medication,
                                    aTNF = pdata_avr$aTNF,
                                    Corticosteroid = pdata_avr$Corticosteroid,
                                    Severity = pdata_avr$Severity,
                                    PC1 = mvals_avr_dmeg_svd$u[,1],
                                    PC2 = mvals_avr_dmeg_svd$u[,2])

mvals_avr_dmeg_svd_df$Severity <- as.character(mvals_avr_dmeg_svd_df$Severity)
mvals_avr_dmeg_svd_df$Severity <- gsub("Active", "CD-active", mvals_avr_dmeg_svd_df$Severity)
mvals_avr_dmeg_svd_df$Severity <- gsub("Remissive", "CD-remissive", mvals_avr_dmeg_svd_df$Severity)

avr_dmeg_svd_plot <- ggplot(mvals_avr_dmeg_svd_df, aes(x = PC1, y = PC2, col = Severity, shape = Medication)) +
  geom_point(size = 4) +
  theme_bw() +
  scale_color_manual(values=c(`CD-remissive` = "#0000ff", `CD-active` = "#ff0000")) +
  xlab(paste0("PC1 (", round(mvals_avr_dmeg_pct_var[1], 1), "%)")) +
  ylab(paste0("PC2 (", round(mvals_avr_dmeg_pct_var[2], 1), "%)")) +
  labs(title = "DMGs",
       subtitle = "Probes associated to the DMGs") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

ggplotly(avr_dmeg_svd_plot)
```

The CD-active with no medication do not appear to cluster separately on either PC1 or PC2, suggesting that medication does not seem to manifest in the DMG-associated CpGs. Note that this does not say that medication has no influence on the DNA methylome - that would be a different question and would likely require a different set of samples.

One of the questions I received was whether we the DMGs are random genes or whether they can actually distinguish CD-active from CD-remissive. Essentially, the question is: what is the predictive value of the DMGs? To answer this question, we can perform some classification analyses. As we have preselected the genes through a meta-analytic approach of linear regressions, we don't need to perform feature selection as we have done so previously. Nonetheless, we will likely need to include some regularization, which are a set of functions that shrink the coefficients of the covariates. Here, we will use ridge regression to ensure that correlated features (CpGs) are shrunk in a reasonably similar fashion. 

```{r classification avr}
require(caret)
require(glmnet)

set.seed(243098)

#caret
avr_train_df <- data.frame(Severity = pdata_avr$Severity,
                           t(mvals_avr_dmeg))

##cross-validation
fitControl <-
  trainControl(
    method = "cv",
    number = 10,
    classProbs = T,
    savePredictions = "all",
    summaryFunction = twoClassSummary
  )

mvals_avr_glmnetcv_caret <- train(
  Severity ~.,
  data = avr_train_df,
  method = "glmnet",
  trControl = fitControl,
  metric = "ROC",
  tuneGrid = expand.grid(alpha = 0, lambda = 10^seq(-3, 3, length = 100))
)

# mvals_avr_glmnetcv_split <- split(mvals_avr_glmnetcv_caret$pred, mvals_avr_glmnetcv_caret$pred$Resample)
# 
# mvals_avr_performance <- do.call(rbind, lapply(names(mvals_avr_glmnetcv_split), function(fold){
#   data.frame(lift(obs ~ Active, data = mvals_avr_glmnetcv_split[[fold]], class = "Active")$data, 
#              fold = fold)
# }))
# 
# ggplot(mvals_avr_performance, aes(x = 1-Sp, y = Sn, col = fold)) +
#   geom_line()
```

We typically observe a ROC of 1, a sensitivity of 1 and a specificity of 1. While this sounds good, it did not surprise me as I had preselected the data based on its discriminative capability albeit using inferential statistics. 

```{r plotting dmgs avr}
ACTvREM_gr <- makeGRangesFromDataFrame(ACTvREM, keep.extra.columns = T, start.field = "pos", end.field = "pos", seqnames.field = "chr")

for(i in 10:length(ACTvREM_ma_sig$Gene)){
  Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgavrDir, paste0(i, "_", ACTvREM_ma_sig$Gene[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[i], tophits_gr = ACTvREM_gr, smooth = T, significance = F, stat = "reduce"))
  dev.off()
}

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(2, "_", ACTvREM_ma_sig$Gene[2], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[2], tophits_gr = ACTvREM_gr[start(ACTvREM_gr) < 1665500 & start(ACTvREM_gr) > 1664500], smooth = T, significance = F))
dev.off()

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(4, "_", ACTvREM_ma_sig$Gene[4], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[4], tophits_gr = ACTvREM_gr[start(ACTvREM_gr) < 38100000], smooth = T, significance = F))
dev.off()

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(dmgDir, paste0(11, "_", ACTvREM_ma_sig$Gene[11], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[11], tophits_gr = ACTvREM_gr[start(ACTvREM_gr) > 58220000], smooth = T, significance = F))
dev.off()
```

Again, we check if the DMGs are associated with any enhancer regions. 

```{r dmg enhancers}
dmg_ACTvREM_enhancers <-lapply(ACTvREM_ma_sig$Gene, function(dmg){
  gene_gr <- ACTvREM_gr[grep(dmg, ACTvREM_gr$UCSC_RefGene_Name),]
  findOverlaps(pchic_gr, gene_gr)
})
ACTvREM_ma_sig[4,]
pchic_gr[unique(queryHits(dmg_ACTvREM_enhancers[[4]]))]

ACTvREM_ma_sig[11,]
pchic_gr[unique(queryHits(dmg_ACTvREM_enhancers[[11]]))]
```

While we find some overlap, the enhancers are only active in erythroblasts.

### DMRs

```{r dmr analysis avr}
dmravrDir <- file.path(dmrDir, "avr")
dir.create(dmravrDir)

dmr_anno_ACTvREM <- cpg.annotate(object = as.matrix(mvals_avr),
                                 datatype = "array",
                                 arraytype = "450K",
                                 what = "M",
                                 analysis.type = "differential",
                                 coef = "SeverityActive",
                                 design = design_avr)
dmr_ACTvREM <- dmrcate(dmr_anno_ACTvREM, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_ACTvREM_gr <- extractRanges(dmr_ACTvREM, genome = "hg19")

write.csv(as.data.frame(dmr_ACTvREM_gr), file.path(dmrDir, "dmrs_ACTvREM.csv"))

dmr_plot(dmr_gr = dmr_ACTvREM_gr[1], meth_data = betas_avr, anno_gr = gmset_anno_gr, meth_groups = pdata_avr$Severity, flanks = 0)
```

Again, we observe some differences, but find it to be less conclusive as compared with the DMRs observed for CD vs non-CD.

## Overlap CD vs non-CD and ACT vs REM

Having identified CD vs non-CD DMGs and CD-active vs CD-remissive DMGs, are there any overlap between the two?

```{r overlap dmgs}
dmg_ol <- merge(CDvNon_CD_ma, ACTvREM_ma, by = "Gene")
colnames(dmg_ol) <- c("Gene", "Brown_pval_cvn", "Fisher_pval_cvn", "Scale_Factor_C_cvn", "DF_cvn", "Brown_padj_cvn", "Brown_pval_avr", "Fisher_pval_avr", "Scale_Factor_C_avr", "DF_avr", "Brown_padj_avr")

dmg_ol$Significance <- "NS"
dmg_ol$Significance[dmg_ol$Brown_padj_cvn <= 0.05 & dmg_ol$Brown_padj_avr <= 0.05] <- "Both"
dmg_ol$Significance[dmg_ol$Brown_padj_cvn <= 0.05 & dmg_ol$Brown_padj_avr > 0.05] <- "CD vs non-CD"
dmg_ol$Significance[dmg_ol$Brown_padj_cvn > 0.05 & dmg_ol$Brown_padj_avr <= 0.05] <- "CD-active vs CD-remissive"
dmg_ol$Significance <- factor(dmg_ol$Significance, levels = c("NS", "CD vs non-CD", "CD-active vs CD-remissive", "Both"))
dmg_ol$euclidian <- sqrt((-log10(dmg_ol$Fisher_pval_cvn))^2+(-log10(dmg_ol$Fisher_pval_avr))^2)
dmg_ol <- dmg_ol[order(dmg_ol$euclidian, decreasing = T),]
dmg_ol$Gene <- as.character(dmg_ol$Gene)
dmg_ol$Gene[grep("C6orf25", as.character(dmg_ol$Gene))] <- "MPIG6B"
dmg_ol$label[dmg_ol$Gene %in% c("MPIG6B")] <- as.character(dmg_ol$Gene[dmg_ol$Gene %in% c("MPIG6B")])

dmg_ol_plot <- ggplot(dmg_ol, aes(x = -log10(dmg_ol$Brown_pval_cvn), y = -log10(dmg_ol$Brown_pval_avr))) +
  geom_point(aes(col = Significance)) +
  geom_text_repel(aes(label = label)) +
  scale_color_manual(values = c(NS = "#D9D5D5", `CD vs non-CD` = "#008000", `CD-active vs CD-remissive` = "#ff00ff", Both = "#6600ff")) +
  labs(title = "Comparison CD vs non-CD with CD-active vs CD-remissive",
       subtitle = "-log10(p-value)") +
  xlab("CD vs non-CD") +
  ylab("CD-active vs CD-remissive") +
  theme_bw() +
  theme(panel.grid.major=element_blank(),
        panel.grid.minor=element_blank())

print(dmg_ol_plot)
```

We find that there is one DMG shared by both, namely `MPIG6B`.

```{r overlapping dmgs}
#MPIG6B (C6orf25)
mpig6b_gr <- makeGRangesFromDataFrame(ACTvREM[grep("C6orf25", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
mpig6b_gr <- mpig6b_gr[start(mpig6b_gr) > 31691000 & start(mpig6b_gr) < 31692000,]

mpig6b_plot <- dmcomp_plot(gene_of_interest = "C6orf25",
                           anno_gr = gmset_anno_gr,
                           tophits1 = CDvNon_CD,
                           tophits2 = ACTvREM,
                           comparison1_name = "CD vs non-CD",
                           comparison2_name = "CD-active vs CD-remissive",
                           meth_data = betas_filtered,
                           meth_groups = pdata_filtered$Severity,
                           ylim = NULL,
                           smooth_m = F)
mpig6b_enlarged_plot <- dmr_plot(dmr_gr = mpig6b_gr, 
                                 meth_data = betas_filtered, 
                                 anno_gr = gmset_anno_gr, 
                                 meth_groups = pdata_filtered$Severity, 
                                 flanks = 0, 
                                 rug = F) +
  scale_color_manual(values=c("#00dae0", "#0000ff", "#ff0000")) +
  theme(axis.title.y = element_blank(),
        legend.pos = "none",
        plot.title = element_blank(),
        plot.subtitle = element_blank())

print(mpig6b_plot)
```

Notably, the differences are opposite with MPIG6B displaying hypomethylation in CD relative to non-CD, but hypermethylation in CD-active relative to CD-remissive. Based on the actual % methylation plot, we observe that the CD-active resembles the non-CD, which is peculiar. I have no clear explanation as to why that would be the case and validations would necessarily be necessary to validate the observations.

## Genes of interest

For some future projects, two genes are of interest to our research group: `GPR146` and `SP140`.

```{r GPR146}
gpr146_gr <- makeGRangesFromDataFrame(ACTvREM[grep("GPR146", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

gpr146_gr[start(gpr146_gr) > 1096000 & start(gpr146_gr) < 1096750,]

# gpr146_plot <- dmcomp_plot(gene_of_interest = "GPR146",
#                            anno_gr = gmset_anno_gr,
#                            tophits1 = CDvNon_CD,
#                            tophits2 = ACTvREM,
#                            comparison1_name = "CD vs non-CD",
#                            comparison2_name = "CD-active vs CD-remissive",
#                            meth_data = betas_filtered,
#                            meth_groups = pdata_filtered$Severity,
#                            ylim = NULL,
#                            smooth_m = F)

gpr146_plot <- dmg_plot(gene_of_interest = "GPR146", 
                        tophits_gr = ACTvREM_gr, 
                        smooth = T, 
                        stat = "identity",
                        significance = F)

gpr146_enlarged_plot <- dmr_plot(dmr_gr = gpr146_gr, 
                                 meth_data = betas_filtered, 
                                 anno_gr = gmset_anno_gr, 
                                 meth_groups = pdata_filtered$Severity, 
                                 flanks = 0, 
                                 rug = F)

Cairo(width = 600, height = 600, type = "pdf", file = "GPR146_cg20603222_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg20603222", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "GPR146_cg26224354_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg26224354", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()
```

While we appear to observe some strange trend, I feel we are looking into the noise as the difference in methylation is less than a 3 percent.

```{r SP140}
sp140_gr <- makeGRangesFromDataFrame(ACTvREM[grep("SP140", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

#sp140_gr <- sp140_gr[start(sp140_gr) >= 231090500,]

# sp140_plot <- dmcomp_plot(gene_of_interest = "SP140",
#                            anno_gr = sp140_gr,
#                            tophits1 = CDvNon_CD,
#                            tophits2 = ACTvREM,
#                            comparison1_name = "CD vs non-CD",
#                            comparison2_name = "CD-active vs CD-remissive",
#                            meth_data = betas_filtered,
#                            meth_groups = pdata_filtered$Severity,
#                            ylim = NULL,
#                            smooth_m = F)

sp140_plot <- dmg_plot(gene_of_interest = "SP140", 
                        tophits_gr = ACTvREM_gr, 
                        smooth = T, 
                        stat = "identity",
                        significance = F)

sp140_enlarged_plot <- dmr_plot(dmr_gr = sp140_gr, 
                                 meth_data = betas_filtered, 
                                 anno_gr = gmset_anno_gr, 
                                 meth_groups = pdata_filtered$Severity, 
                                 flanks = 0, 
                                 rug = F)

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg03887528_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg03887528", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg05564251_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg05564251", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg04579254_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg04579254", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()
```

For SP140, we hardly see any difference

## Public data

As DNA methylation is a feature specific to a particular cell-type and CD14+ monocytes consist of different subpopulations (classical, non-classical and intermediate), it stands to reason that the observed differences in methylation could be the result of changes in the underlying monocyte populations. While this would require a different experiment to properly investigate this, we can approximate an answer by investigating by comparing our data with methylation data on the monocyte subsets as published by [Zawada et al. 2016](https://www.ncbi.nlm.nih.gov/pubmed/27018948).

```{r setup public data}
publicDir <- file.path(outputDir, "public")
dir.create(publicDir)
```

### Methyl-Seq: GSE73788

```{r setup methylseq}
mseqDir <- file.path(publicDir, "methylseq")
dir.create(mseqDir)
```

```{r gse73788 import and processing}
gse73788 <- read.csv(file.path("data", "methylseq", "GSE73788_Mo1_Mo2_Mo3_merged.csv"), stringsAsFactors = F)
gse73788$seqnames <- gsub("(chr[0-9XYM]+):[0-9]+-[0-9]+", "\\1", gse73788$Position)
gse73788$start <- as.numeric(gsub("chr[0-9XYM]+:([0-9]+)-[0-9]+", "\\1", gse73788$Position))
gse73788$end <- as.numeric(gsub("chr[0-9XYM]+:[0-9]+-([0-9]+)", "\\1", gse73788$Position))
gse73788_gr <- makeGRangesFromDataFrame(gse73788, keep.extra.columns = T)

gse73788_meth <- data.frame(gse73788[, c("Mo1", "Mo2", "Mo3")])
gse73788_meth_l2i <- -log2(gse73788_meth+1)
gse73788_groups <- c("Classical", "Intermediate", "Non-classical")
```

```{r CD14 and CD16}
gse73788_gr[grep("CD14", gse73788_gr$Gene),]
gse73788_gr[grep("CD16", gse73788_gr$Gene),]
```

```{r visualize}
#dmgs_sig <- union(ACTvREM_ma_sig$Gene, CDvNon_CD_ma_sig$Gene)
dmgs_sig <- c("CD14", "CD16")

for(i in 41:length(dmgs_sig)){
  goi_gr <- range(gmset_anno_gr[grep(paste0("(^|;)", dmgs_sig[i], "($|;)"), gmset_anno_gr$UCSC_RefGene_Name),], ignore.strand = T)
  
  Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(mseqDir, paste0(dmgs_sig[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(dmrma_plot(roi_gr = goi_gr,
                   ownmeth_data = betas_filtered,
                   ownmeth_anno_gr = gmset_anno_gr,
                   ownmeth_groups = pdata_filtered$Severity,
                   pubmeth_data = gse73788_meth_l2i,
                   pubmeth_anno_gr = gse73788_gr,
                   pubmeth_groups = gse73788_groups))
  dev.off()
}
```

In general it appears as though `Mo1` has more methylation than `Mo2` or `Mo3`. Overlaying the methylation by Zawada  onto the DNA methylation data at hand seems to corroborate the broad trends, but lacks the granularity. 

The lack in overlap appears to be caused by two issues in particular:
* The difference in technology: Zawada et al. used HpaII to identify unmethylated CpGs and did not use a MspI to measure all CG regions, making the inference on percentage methylation difficult.
* The lack of biological replicates. Zawada et al. describe the pooling of monocytes from 7 donors and sequencing them in one go. Based on the available data, the samples were not indexed and hence the observed signal is at it stands.

# Tables

```{r sample information}
require(dplyr)

#Age
pdata_filtered %>% 
  data.frame() %>%
  select(Age, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(Age), stdev = sd(Age))

#CRP
pdata_filtered %>% 
  data.frame() %>%
  select(CRP, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(CRP, na.rm = T), stdev = sd(CRP, na.rm = T))

#HBI
pdata_filtered %>% 
  data.frame() %>%
  select(HBI, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(HBI, na.rm = T), stdev = sd(HBI, na.rm = T))

#Medication
pdata_filtered %>% 
  data.frame() %>%
  select(Medication, Severity) %>%
  group_by(Severity) %>%
  table()

#anti-TNF
pdata_filtered %>% 
  data.frame() %>%
  select(aTNF, Severity) %>%
  group_by(Severity) %>%
  table()

#Corticosteroid
pdata_filtered %>% 
  data.frame() %>%
  select(Corticosteroid, Severity) %>%
  group_by(Severity) %>%
  table()

#Thiopurine
pdata_filtered %>% 
  data.frame() %>%
  select(Thiopurine, Severity) %>%
  group_by(Severity) %>%
  table()
```

## Future research

It might make sense to perform a well-powered analysis of some of the interesting features we highlighted in this study. For example, the CD vs non-CD DMR could be validated through targeted bisulfite sequencing. On the question of sample size, we could perform a post-hoc power calculation. Note that post-hoc power calculations are not necessarily useful given the fact that statistical power is positively correlated with p-values, see this [blog](http://daniellakens.blogspot.com/2014/12/observed-power-and-what-to-do-if-your.html).

```{r dmr cvn power calculation}
powerDir <- file.path(outputDir, "power_analysis")
dir.create(powerDir)

library(pwr)

cohend <- function(m1, m2, s1, s2){
  (m1 - m2)/sqrt(((s1^2) + (s2^2))/2) 
} 

dmr_cpgs <- names(subsetByOverlaps(gmset_anno_gr, dmr_CDvNon_CD_gr[1]))
CDvNon_CD[dmr_cpgs,]
dmr_betas <- melt(cbind(betas_filtered[dmr_cpgs,], dmr_cpgs))
dmr_betas$Status <- pdata_filtered[dmr_betas$variable, "Status1"]
colnames(dmr_betas) <- c("CpG", "ID", "Beta", "Status")

dmr_summarized <- dmr_betas %>% 
  group_by(Status, CpG) %>%
  summarise(
    mu = mean(Beta),
    sdev = sd(Beta)
  ) %>% 
  arrange(CpG)
  #summarise_all(funs(trimws(paste(., collapse = ''))))

dmr_summarized <- split(dmr_summarized, dmr_summarized$Status)
dmr_summarized <- merge(dmr_summarized$CD, dmr_summarized$Non_CD, by = "CpG")
rownames(dmr_summarized) <- dmr_summarized$CpG

dmr_summarized$N <- do.call(c, lapply(rownames(dmr_summarized), function(cpg){
  cohd <- cohend(m1 = dmr_summarized[cpg, "mu.x"], m2 = dmr_summarized[cpg, "mu.y"], s1 = dmr_summarized[cpg, "sdev.x"], s2 = dmr_summarized[cpg, "sdev.y"])
  pwr.t.test(
    n = NULL,                  # Observations in _each_ group
    d = cohd,           
    sig.level = 0.001,          # Type I probability
    power = 0.80,              # 1 minus Type II probability
    type = "two.sample",       # Change for one- or two-sample
    alternative = "two.sided"
  )$n
}))

pwr.t.test(
  n = NULL,                  # Observations in _each_ group
  d = 2.456725,           
  sig.level = 0.001,          # Type I probability
  power = 0.80,              # 1 minus Type II probability
  type = "two.sample",       # Change for one- or two-sample
  alternative = "two.sided"
)

dmr_summarized <- dmr_summarized[dmr_cpgs,]
write.csv(dmr_summarized, file.path(powerDir, "power_dmr_cpgs.csv"))
```

For the actual experiments, we could use primers designed using the following primers.

```{r primer}
fig1propDir <- file.path(proposalDir, "fig1")
dir.create(fig1propDir)

tab1propDir <- file.path(proposalDir, "table1")
dir.create(tab1propDir)

require(GenomicRanges)
amplicons <- data.frame(chr = rep("chr7", 9), 
                        start = c(start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 744-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 733-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 463-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 744-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 246-1),
                        end = c(start(dmr_CDvNon_CD_gr[1]) + 479,
                                start(dmr_CDvNon_CD_gr[1]) + 475,
                                start(dmr_CDvNon_CD_gr[1]) + 476,
                                start(dmr_CDvNon_CD_gr[1]) + 471,
                                start(dmr_CDvNon_CD_gr[1]) + 976,
                                start(dmr_CDvNon_CD_gr[1]) + 976,
                                start(dmr_CDvNon_CD_gr[1]) + 611,
                                start(dmr_CDvNon_CD_gr[1]) + 977,
                                start(dmr_CDvNon_CD_gr[1]) + 475),
                        L_tm = c(54.95, 54.95, 54.95, 54.95, 54.71, 52.14, 57.38, 54.71, 56.22),
                        R_tm = c(58.17 , 59.30, 58.99, 59.92, 54.90, 54.90, 57.88, 56.23, 59.30),
                        L_GC = c(40.00, 40.00, 40.00, 40.00, 53.57, 40.00, 60.00, 53.57, 38.46),
                        R_GC = c(60.00, 56.00, 56.00, 57.69, 58.33, 58.33, 53.85, 56.00, 56.00),
                        L_seq = c("TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "GAGGAGTTGTTATTTTTATTTTATTTAA", "GTAGAGTGTTTTAGAAATTTTTTTT", "GGGAGTTGTGTATATTTAGGTTGAG", "GAGGAGTTGTTATTTTTATTTTATTTAA", "TTTAGAGAATTTTTGGATTTGTAAGT"),
                        R_seq = c("AAACTATACCAAAAAAACCCTCAAC", "TATACCAAAAAAACCCTCAACCTAA", "CTATACCAAAAAAACCCTCAACCTA", "CCAAAAAAACCCTCAACCTAAATATA", "AAAACTATATCTCCCTTACCACAC", "AAAACTATATCTCCCTTACCACAC", "CTATAAAAAACAATACAACCCTCTCC", "AAAAACTATATCTCCCTTACCACAC", "TATACCAAAAAAACCCTCAACCTAA"),
                        tm = c(64.7, 64.8, 64.8, 64.5, 64.7, 64.6, 64.6, 64.7, 64.8),
                        prod_nCpGs = c(11, 11, 11, 11, 11, 10, 6, 11, 11),
                        ymin = seq(0, 4, by = 0.5),
                        ymax = seq(0.5, 4.5, by = 0.5))

amplicons$name <- paste0("amp", 1:9)
amplicons$prod_length <- amplicons$end - amplicons$start
amplicons$L_bps <- nchar(as.character(amplicons$L_seq))
amplicons$R_bps <- nchar(as.character(amplicons$R_seq))

amplicons <- amplicons[,c("chr", "start", "end", "name", "L_tm", "R_tm", "tm", "L_GC", "R_GC", "L_bps", "R_bps", "prod_nCpGs", "prod_length", "L_seq", "R_seq", "ymin", "ymax")]
write.csv(amplicons, file.path(tab1propDir, "table1.csv"))

dmr1 <- dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Status1, flanks = 0) + theme(legend.pos = "top", axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())
amps <- ggplot(amplicons, aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + 
  geom_rect(alpha = 0.25, col = "black") + 
  geom_text(aes(x=start+(end-start)/2, y=ymin+(ymax-ymin)/2, label=name), size = 4) +
  #facet_wrap(~name, ncol = 1, strip.position = "left") +
  theme_bw() +
  xlim(start(dmr_CDvNon_CD_gr[1]), end(dmr_CDvNon_CD_gr[1])) +
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.text = element_text(size = 12), 
        axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 750, height = 1000, type = "png", file = file.path(fig1propDir, "fig1.png"), bg = "white", units = "px", dpi = 120)
print(ggarrange(dmr1, amps, ncol = 1, nrow = 2, align = "v"))
dev.off()

Cairo(width = 1200, height = 1600, type = "pdf", file = file.path(fig1propDir, "fig1.pdf"), bg = "white", units = "px", dpi = 120)
print(ggarrange(dmr1, amps, ncol = 1, nrow = 2, align = "v"))
dev.off()
```

# Figures

```{r Figure 1}
require(ggpubr)
require(Cairo)

fig1Dir <- file.path(manuscriptDir, "fig1")
dir.create(fig1Dir)

#A
Cairo(width = 2250, height = 1000, type = "pdf", file = file.path(fig1Dir, "fig1a.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap(t(betas_t50_cvn_dm), annotation_row = hm_anno_t50_cvn, show_rownames = F)
dev.off()

#B
CDvNon_CD_ma_th <- CDvNon_CD_ma[1:50,]
CDvNon_CD_ma_th$Gene <- factor(CDvNon_CD_ma_th$Gene, levels = CDvNon_CD_ma_th$Gene)
CDvNon_CD_ma_th$Significance <- "Non-significant"
CDvNon_CD_ma_th$Significance[CDvNon_CD_ma_th$Brown_padj<0.05] <- "Significant"

Cairo(width = 2250, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1b.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(CDvNon_CD_ma_th, aes(x = Gene, y = -log10(Brown_pval), fill = Significance)) +
  geom_bar(stat = "identity") +
  ylab("-log10(p-value)") +
  theme_bw() +
  #coord_flip() +
  scale_fill_manual(values=c(Significant = "#000000", `Non-significant` = "#d3d3d3")) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1),
        axis.title.x = element_blank())
dev.off()
```

```{r Figure 2}
fig2Dir <- file.path(manuscriptDir, "fig2")
dir.create(fig2Dir)

#A
Cairo(width = 1250, height = 1500, type = "pdf", file = file.path(fig2Dir, "fig2a.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap(betas_t50_avr_dm, annotation_col = hm_anno_t50_avr, show_colnames = F)
dev.off()

#B
ACTvREM_ma_th <- ACTvREM_ma[1:50,]
ACTvREM_ma_th$Gene <- factor(ACTvREM_ma_th$Gene, levels = rev(ACTvREM_ma_th$Gene))
ACTvREM_ma_th$Significance <- "Non-significant"
ACTvREM_ma_th$Significance[ACTvREM_ma_th$Brown_padj<0.05] <- "Significant"

Cairo(width = 1000, height = 1500, type = "pdf", file = file.path(fig2Dir, "fig2b.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(ACTvREM_ma_th, aes(x = Gene, y = -log10(Brown_pval), fill = Significance)) +
  geom_bar(stat = "identity") +
  ylab("-log10(p-value)") +
  theme_bw() +
  coord_flip() +
  scale_fill_manual(values=c(Significant = "#000000", `Non-significant` = "#d3d3d3")) +
  theme(panel.grid.major = element_blank(),
        panel.grid.minor = element_blank(),
        #axis.text.y = element_text(angle = 90, hjust = 1),
        axis.title.y = element_blank())
dev.off()

#D
Cairo(width = 1500, height = 1200, type = "pdf", file = file.path(fig2Dir, "fig2d.pdf"), bg = "white", units = "px", dpi = 120)
print(avr_dmeg_svd_plot)
dev.off()
```

```{r Figure 3}
fig3Dir <- file.path(manuscriptDir, "fig3")
dir.create(fig3Dir)

#A
Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(fig3Dir, "fig3a.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_ol_plot)
dev.off()

#B
Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(fig3Dir, "fig3b1.pdf"), bg = "white", units = "px", dpi = 120)
print(mpig6b_plot)
dev.off()

Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(fig3Dir, "fig3b2.pdf"), bg = "white", units = "px", dpi = 120)
print(mpig6b_enlarged_plot)
dev.off()
```

```{r Figure S1}
figS1Dir <- file.path(manuscriptDir, "figS1")
dir.create(figS1Dir)

figS1_df <- mvals_filtered_svd_df
figS1_df$Severity <- as.character(figS1_df$Severity)
figS1_df$Severity <- gsub("Active", "CD-active", figS1_df$Severity)
figS1_df$Severity <- gsub("Remissive", "CD-remissive", figS1_df$Severity)
figS1_df$Severity <- gsub("Non_CD", "non-CD", figS1_df$Severity)
figS1_df$Severity <- factor(figS1_df$Severity, levels = c("non-CD", "CD-remissive", "CD-active"))

pct_var <- mvals_filtered_svd$d/sum(mvals_filtered_svd$d)*100

figS1_plot <- ggplot(figS1_df, aes(x = PC1, y = PC2, fill = Severity)) +
  geom_point(shape = 21, size = 4) +
  theme_bw() +
  scale_fill_manual(values=c(`non-CD` = "#00dae0",`CD-remissive` = "#0000ff", `CD-active` = "#ff0000")) +
  xlab(paste0("PC1 (", round(pct_var[1], 1), "%)")) +
  ylab(paste0("PC2 (", round(pct_var[2], 1), "%)")) +
  labs(title = "Principal component analysis") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())

Cairo(width = 1000, height = 750, type = "pdf", file = file.path(figS1Dir, "figS1.pdf"), bg = "white", units = "px", dpi = 120)
print(figS1_plot)
dev.off()
```

```{r Figure S2}
figS2Dir <- file.path(manuscriptDir, "figS2")
dir.create(figS2Dir)

ggplot(data.frame(CDvNon_CD), aes(x = P.Value)) +
  geom_bar() +
  theme_bw()
```

```{r Figure S3}
figS3Dir <- file.path(manuscriptDir, "figS1")
dir.create(figS3Dir)

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(figS3Dir, "figS3.pdf"), bg = "white", units = "px", dpi = 120)
#dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity, flanks = 0)
dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Status1, flanks = 0, title = "CD vs non-CD DMR 1") + 
  scale_color_manual(values=c(`Non_CD` = "#00dae0",`CD` = "#ff00ff"))
dev.off()
```

