---
title: "DM analysis PRJ0000002_CDMON"
author: "Andrew Li Yim"
date: "November 4, 2015"
output: html_document
editor_options: 
  chunk_output_type: console
---

In this report we will perform the differential methylation analyses on CD14+ monocytes derived from CD and non-CD patients. The CD patients can be further subdivided into active and remissive CD.

```{r setup, echo = F}
require(pheatmap)
require(minfi)
require(aggregation)
require(EmpiricalBrownsMethod)

samplesDir <- file.path("data", "samples")
commondataDir <- file.path("/home", "ayliyim", "archive", "common_data")

manuscriptDir <- file.path("docs", "manuscript", "200106")
dir.create(manuscriptDir)
proposalDir <- file.path("docs", "proposal")
dir.create(proposalDir)

samplesheet <- read.metharray.sheet(base = samplesDir, pattern = "samplesheet_450k_PRJ0000002_CDMON_V6")
samplesheet$Basename <- samplesheet$Path
samplesheet$Sex <- "F"

rgset <- read.metharray.exp(targets = samplesheet)
```

```{r bespoke functions}
source(file.path("src", "dmr_plot.R"))
source(file.path("src", "dmg_plot.R"))
source(file.path("src", "dmcomp_plot.R"))
```

Date the output directory.

```{r output directory}
outputDir <- file.path("output", "01_DM_450k_analyses", "200106")
dir.create(outputDir)
```

```{r sample information}
require(dplyr)

#Age
samplesheet %>% 
  select(Age, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(Age), stdev = sd(Age))

#CRP
samplesheet %>% 
  select(CRP, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(CRP, na.rm = T), stdev = sd(CRP, na.rm = T))

#HBI
samplesheet %>% 
  select(HBI, Severity) %>%
  group_by(Severity) %>%
  summarize(mean = mean(HBI, na.rm = T), stdev = sd(HBI, na.rm = T))

#anti-TNF
samplesheet %>% 
  select(aTNF, Severity) %>%
  group_by(Severity) %>%
  summarize(n = n(), percentage = sd(aTNF, na.rm = T))

#Corticosteroid
samplesheet %>% 
  select(Corticosteroid, Severity) %>%
  group_by(Severity) %>%
  summarize(n = n())
```


```{r Quality control, echo = FALSE}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)

# require(MethylAid)
# methylaid_summarized <- summarize(samplesheet)
# visualize(methylaid_summarized)
# 
# require(shinyMethyl)
# shinymethyl_summarized <- shinySummarize(rgset)
# runShinyMethyl(shinymethyl_summarized)
# 
# require(ewastools)
# ewastools_samples <- read_idats(samplesheet$Basename)
# ewastools_betas <- dont_normalize(ewastools_samples)
# ewastools_snps <- call_genotypes(ewastools_betas[ewastools_samples$manifest[probe_type == 'rs', index], ], learn = F)
# check_snp_agreement(genotypes = ewastools_snps$gamma, weights = 1 - ewastools_snps$outliers, donor_ids = samplesheet$Donor_ID, sample_ids = samplesheet$Sample_ID)
```

According to MethylAid, all samples pass the quality control. Additionally, the ewastools package was used to confirm that the samples were obtained from the same origin.
As the sample pertains peripheral blood, we can utilize the estimateCellCounts function from minfi to see whether we observe any differences in estimated cell populations between the various groups.

```{r Blood cell distribution}
require(FlowSorted.Blood.450k)
require(ggplot2)
require(Cairo)

cellCounts <- estimateCellCounts(rgSet = rgset, 
                                 compositeCellType = "Blood",
                                 processMethod = "auto", 
                                 probeSelect = "auto")
cellCounts <- data.frame(cellCounts, severity = pData(rgset)$Severity, array = rownames(cellCounts))

cellCounts_melt <- reshape2::melt(cellCounts, id.vars = c("severity", "array"), id.vals = c("Response"))

cellCounts_plot <- ggplot(cellCounts_melt, aes(x = severity, y = value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter() +
  facet_wrap(~variable) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 1000, height = 1000, type = "png", file = file.path(edaDir, "estimatedCellCounts.png"), bg = "white", units = "px", dpi = 120)
print(cellCounts_plot)
dev.off()
```

The cell distribution accurately guesses that the samples pertain monocytes. There appears to be one sample however that appears more CD4T-like.

```{r CD4T samples}
require(dplyr)

CD4T_array_ID <- cellCounts_melt %>% 
  filter(variable == "Mono" & value < 0.25) %>% 
  select(array)

pData(rgset)[CD4T_array_ID$array, "Sample_ID"]
```

We will next investigate whether this particular sample is an outlier in the PCA as well. Before that, we will remove the regular outliers.

```{r Preprocessing}
gmset <- preprocessFunnorm(rgSet = rgset)
gmset_nosnpcpg <- gmset[with(getSnpInfo(gmset), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
gmset_nosnpsbe <- gmset_nosnpcpg[with(getSnpInfo(gmset_nosnpcpg), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
gmset_nosex <- gmset_nosnpsbe[-which(getAnnotation(gmset_nosnpsbe)$chr %in% c("chrX", "chrY")),]

pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChip450k", "Non-specific-probes-Illumina450k.csv"), stringsAsFactors = F, header = T)[,1]
gmset_filtered <- gmset_nosex[!names(gmset_nosex) %in% pmprobes,]

mvals <- getM(gmset_filtered)

gmset_filtered <- gmset_filtered[-which(abs(mvals) == Inf, arr.ind = T)[,1],]

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

Are there any sex discrepancies?

```{r Annotated and predicted sex}
require(ggplot2)
require(dplyr)
require(ggrepel)

pData(gmset_filtered) %>% 
  data.frame %>% 
  select(Sample_ID, Sex, predictedSex, xMed, yMed) %>%
  mutate(Sample_label = ifelse(Sex == predictedSex, "", Sample_ID)) %>% 
  ggplot(aes(x = xMed, y = yMed, col = Sex, label = Sample_label)) +
  geom_text_repel(show.legend = F) +
  geom_point(size = 4) +
  theme_bw() +
  coord_fixed() +
  ggtitle("Sex chromosome probe median intensities") +
  xlab("chrX") +
  ylab("chrY") +
  theme(text = element_text(size=20))
```

```{r PCA}
require(plotly)

mvals_dm <- mvals - rowMeans(mvals)
mvals_svd <- svd(t(mvals_dm))

mvals_svd_df <- data.frame(Sample_ID = pData(gmset_filtered)$Sample_ID,
                           Origin = pData(gmset_filtered)$Donor_ID,
                           Age = pData(gmset_filtered)$Age,
                           Severity = pData(gmset_filtered)$Severity,
                           Sex = pData(gmset_filtered)$predictedSex,
                           PC1 = mvals_svd$u[,1],
                           PC2 = mvals_svd$u[,2])

mvals_svd_ggplotly <- ggplot(mvals_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Severity = Severity, Age = Age, Sex = Sex)) +
  geom_point(aes(col = Origin)) +
  theme_bw() +
  theme(text = element_text(size=20))

ggplotly(mvals_svd_ggplotly)

mvals_svd_plotobj <- ggplot(mvals_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Severity = Severity, Age = Age, Sex = Sex)) +
  geom_point(aes(col = Origin)) +
  theme_bw() +
  theme(text = element_text(size=20))
```

We again observe an outlier on PC1.

```{r PCA outlier}
mvals_svd_df[which.max(mvals_svd_df$PC1),]
```

This outlier appears to be HC1 again. Due to its outlier status, we will therefore remove HC1 from further analyses.

```{r HC1 removal}
gmset_filtered <- gmset_filtered[, pData(gmset_filtered)$Sample_ID != "HC1"]
pData(gmset_filtered)$Severity <- factor(pData(gmset_filtered)$Severity, levels = c("Non_CD", "Remissive", "Active"))

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

There are two samples derived from the same donor: CDACT-1a and CDACT-1b. As there is no clear way of choosing one over the other, we will take the mean.

```{r CDACT1 aggregation}
mvals_filtered <- data.frame(CDACT1 = rowMeans(mvals[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), mvals[, 3:ncol(mvals)])
betas_filtered <- data.frame(CDACT1 = rowMeans(betas[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), betas[, 3:ncol(betas)])

pdata_filtered <- pData(gmset_filtered)[-1,]
colnames(mvals_filtered) <- colnames(betas_filtered) <- rownames(pdata_filtered)
```

```{r pca filtered}
mvals_filtered_dm <- mvals_filtered - rowMeans(mvals_filtered)
mvals_filtered_svd <- svd(t(mvals_filtered_dm))

mvals_filtered_svd_df <- data.frame(Sample_ID = pdata_filtered$Sample_ID,
                                    Origin = pdata_filtered$Donor_ID,
                                    Age = pdata_filtered$Age,
                                    Severity = pdata_filtered$Severity,
                                    PC1 = mvals_filtered_svd$u[,1],
                                    PC2 = mvals_filtered_svd$u[,2])

mvals_filtered_svd_plotobj <- ggplot(mvals_filtered_svd_df, aes(x = PC1, y = PC2, fill = Severity)) +
  geom_point(shape = 21, size = 4) +
  theme_bw() +
  labs(title = "Principal component analysis") +
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank())
```

Prepare annotation for the eventual interpretation

```{r Annotation preparation}
gmset_anno <- getAnnotation(gmset_filtered)
gmset_anno_gr <- makeGRangesFromDataFrame(gmset_anno, keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
```

```{r Output directories}
dmpDir <- file.path(outputDir, "dmps")
dir.create(dmpDir)
dmrDir <- file.path(outputDir, "dmrs")
dir.create(dmrDir)
dmgDir <- file.path(outputDir, "dmgs")
dir.create(dmgDir)
```

#CD vs non-CD

First we compare CD with non-CD.

```{r DM analyses preparation cvn}
require(limma)

pdata_filtered$Status1 <- factor(pdata_filtered$Status1, levels = c("Non_CD", "CD"))

design_cvn <- model.matrix(~Status1 + Age, data = pdata_filtered)

mvals_lmfit_cvn <- lmFit(mvals_filtered, design = design_cvn)
betas_lmfit_cvn <- lmFit(betas_filtered, design = design_cvn)

mvals_ebayes <- eBayes(mvals_lmfit_cvn)
betas_ebayes <- eBayes(betas_lmfit_cvn)
```

```{r DMP analysis}
require(NDlib)

CDvNon_CD <- topTable(mvals_ebayes, coef = "Status1CD", number = "Inf", adjust.method = "BH")
CDvNon_CD <- cbind(CDvNon_CD, Beta = coefficients(betas_ebayes)[rownames(CDvNon_CD), "Status1CD"], gmset_anno[rownames(CDvNon_CD), ])

write.csv(CDvNon_CD, file.path(dmpDir, "CDvNon_CD.csv"))

for(i in 1:10){
  Cairo(width = 600, height = 800, type = "pdf", file = file.path(dmpDir, paste0(i, "_", rownames(CDvNon_CD)[i], "_CDvNon_CD.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(CDvNon_CD)[i], betas = betas_filtered, factor_interest = pdata_filtered$Status1, enlarged = F, type = "SE", legend = F) + 
          theme(panel.grid.major=element_blank(),
                panel.grid.minor=element_blank()))
  dev.off()
}
```

```{r Heatmap cvn}
hm_anno_t50_cvn <- data.frame(Status = pdata_filtered$Status1, row.names = rownames(pdata_filtered))

betas_t50_cvn <- betas_filtered[rownames(CDvNon_CD[1:50,]),]
betas_t50_cvn_dm <- betas_t50_cvn-rowMeans(betas_t50_cvn)

pheatmap(betas_t50_cvn_dm, annotation_col = hm_anno_t50_cvn)
```

```{r Surrounding CpGs cvn}
#cg17206772
cg17206772_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[1])
cg17206772_area_gr <- gmset_anno_gr[(cg17206772_area-5):(cg17206772_area+5),]

dmr_plot(dmr_gr = range(cg17206772_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)

#cg16921727
cg16921727_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[2])
cg16921727_area_gr <- gmset_anno_gr[(cg16921727_area-5):(cg16921727_area+5),]

dmr_plot(dmr_gr = range(cg16921727_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)
```

```{r Differentially methylated genes cvn}
CDvNon_CD_genes <- lapply(strsplit(CDvNon_CD$UCSC_RefGene_Name, ";"), unique)
names(CDvNon_CD_genes) <- rownames(CDvNon_CD)
CDvNon_CD_genes_nz <- CDvNon_CD_genes[which(lapply(CDvNon_CD_genes, length) != 0)]

CDvNon_CD_genes_nz_df <- data.frame(CpG = gsub("(cg[0-9]{8})[0-9]", "\\1", names(unlist(CDvNon_CD_genes_nz))), 
                                    Gene = unlist(CDvNon_CD_genes_nz))
CDvNon_CD_genes_nz_df$pvalue <- CDvNon_CD[CDvNon_CD_genes_nz_df$CpG, "P.Value"]

CDvNon_CD_genes_nz_list <- split(CDvNon_CD_genes_nz_df, CDvNon_CD_genes_nz_df$Gene)

CDvNon_CD_fisher <- lapply(CDvNon_CD_genes_nz_list, function(gene){
  aggregation::fisher(gene$pvalue)
})

rm(CDvNon_CD_genes_nz_list)

CDvNon_CD_ma <- data.frame(Gene = names(CDvNon_CD_fisher),
                           Fisher_pval = unlist(CDvNon_CD_fisher))
CDvNon_CD_ma$padj <- p.adjust(CDvNon_CD_ma$Fisher_pval)
CDvNon_CD_ma <- CDvNon_CD_ma[order(CDvNon_CD_ma$Fisher_pval),]
CDvNon_CD_ma_sig <- CDvNon_CD_ma[which(CDvNon_CD_ma$padj<0.05),]

write.csv(CDvNon_CD_ma, file.path(dmgDir, "CDvNon_CD.csv"))
```

```{r plotting dmgs cvn}
CDvNon_CD_gr <- makeGRangesFromDataFrame(CDvNon_CD, keep.extra.columns = T, start.field = "pos", end.field = "pos", seqnames.field = "chr")

for(i in 1:length(CDvNon_CD_ma_sig$Gene)){
  Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(dmgDir, paste0(i, "_", CDvNon_CD_ma_sig$Gene[i], "_CDvnonCD.pdf")), bg = "white", units = "px", dpi = 120)
  print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[i], tophits_gr = CDvNon_CD_gr, smooth = T, significance = F))
  dev.off()
}
```

```{r gene set analyses cvn}
require(DOSE)
require(Biobase)
require(org.Hs.eg.db)
require(clusterProfiler)

CDvNon_CD_ma_sig$Entrez <- mapIds(Homo.sapiens, 
                                  keys = as.character(CDvNon_CD_ma_sig$Gene),
                                  column = "ENTREZID",
                                  keytype = "SYMBOL",
                                  multiVals = "first")
CDvNon_CD_ma_sig["C6orf25", "Entrez"] <- 80739
CDvNon_CD_ma_sig["C22orf26", "Entrez"] <- 55267
CDvNon_CD_ma_sig["PNMAL2", "Entrez"] <- 57469
CDvNon_CD_ma_sig["C10orf26", "Entrez"] <- 54838

CDvNon_CD_ma_sig_kegg <- enrichKEGG(gene = CDvNon_CD_ma_sig$Entrez,
                                    organism = 'hsa')
```

```{r DMR analysis cvn}
dmrcvnDir <- file.path(dmrDir, "cvn")
dir.create(dmrcvnDir)

require(DMRcate)
dmr_anno_CDvNon_CD <- cpg.annotate(object = as.matrix(mvals_filtered),
                                   datatype = "array",
                                   arraytype = "450K",
                                   what = "M",
                                   analysis.type = "differential",
                                   coef = "Status1CD",
                                   design = design_cvn)
dmr_CDvNon_CD <- dmrcate(dmr_anno_CDvNon_CD, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_CDvNon_CD_gr <- extractRanges(dmr_CDvNon_CD, genome = "hg19")

write.csv(as.data.frame(dmr_CDvNon_CD_gr), file.path(dmrDir, "dmrs_CDvNon_CD.csv"))

dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity, flanks = 0)

dmr_cpgs <- names(subsetByOverlaps(gmset_anno_gr, dmr_CDvNon_CD_gr[1]))
CDvNon_CD[dmr_cpgs,]
dmr_betas <- melt(cbind(betas_filtered[dmr_cpgs,], dmr_cpgs))
dmr_betas$Status <- pdata_filtered[dmr_betas$variable, "Status1"]
colnames(dmr_betas) <- c("CpG", "ID", "Beta", "Status")
```

```{r Power calculation}
powerDir <- file.path(outputDir, "power_analysis")
dir.create(powerDir)

library(pwr)

cohend <- function(m1, m2, s1, s2){
  (m1 - m2)/sqrt(((s1^2) + (s2^2))/2) 
} 

dmr_summarized <- dmr_betas %>% 
  group_by(Status, CpG) %>%
  summarise(
    mu = mean(Beta),
    sdev = sd(Beta)
  ) %>% 
  arrange(CpG)
  #summarise_all(funs(trimws(paste(., collapse = ''))))

dmr_summarized <- split(dmr_summarized, dmr_summarized$Status)
dmr_summarized <- merge(dmr_summarized$CD, dmr_summarized$Non_CD, by = "CpG")
rownames(dmr_summarized) <- dmr_summarized$CpG

dmr_summarized$N <- do.call(c, lapply(rownames(dmr_summarized), function(cpg){
  cohd <- cohend(m1 = dmr_summarized[cpg, "mu.x"], m2 = dmr_summarized[cpg, "mu.y"], s1 = dmr_summarized[cpg, "sdev.x"], s2 = dmr_summarized[cpg, "sdev.y"])
  pwr.t.test(
    n = NULL,                  # Observations in _each_ group
    d = cohd,           
    sig.level = 0.001,          # Type I probability
    power = 0.80,              # 1 minus Type II probability
    type = "two.sample",       # Change for one- or two-sample
    alternative = "two.sided"
  )$n
}))

pwr.t.test(
       n = NULL,                  # Observations in _each_ group
       d = 2.456725,           
       sig.level = 0.001,          # Type I probability
       power = 0.80,              # 1 minus Type II probability
       type = "two.sample",       # Change for one- or two-sample
       alternative = "two.sided"
       )

dmr_summarized <- dmr_summarized[dmr_cpgs,]
write.csv(dmr_summarized, file.path(powerDir, "power_dmr_cpgs.csv"))
```

We observe one DMR that we have observed previously: chr7:51538650-51539678. In this region, CD samples are hypermethylated relative to non-CD samples. Unfortunately, Illumina has no annotation for the probes within this region. One possible reason is that this region falls within an enhancer. We can use the Javierre et al. 2016 data to investigate this possibility. 

```{r Enhancers pcHiC}
pchic <- read.csv("~/archive/common_data/enhancers/javierre2016/ActivePromoterEnhancerLinks_hg19.tsv", sep = "\t")
pchic_gr <- makeGRangesFromDataFrame(pchic, keep.extra.columns = T, seqnames.field = "oeChr", start.field = "oeSt", end.field = "oeEnd")

findOverlaps(pchic_gr, dmr_CDvNon_CD_gr[1])
```
No overlaps were found, but Javierre et al. might not be the only source to look at. We will next Try Ensembl Regulatory. As Ensembl Regulatory uses GRCh38, we will need to liftover the coordinates.

```{r Promoters Ensembl Regulatory}
require(rtracklayer)
hg19togrch38 <- import.chain("~/archive/common_data/liftover_chains/hg19ToHg38.over.chain")
dmr_CDvNon_CD_gr_grch38 <- liftOver(dmr_CDvNon_CD_gr[1], hg19togrch38)
```

##ACT vs REM

```{r DM analyses setup avr}
dmpavrDir <- file.path(dmpDir, "avr")
dir.create(dmpavrDir)

pdata_avr <- pdata_filtered[pdata_filtered$Status1 == "CD", ]
pdata_avr$Severity <- factor(pdata_avr$Severity, c("Remissive", "Active"))
mvals_avr <- mvals_filtered[,rownames(pdata_avr)]
betas_avr <- betas_filtered[,rownames(pdata_avr)]

design_avr <- model.matrix(~Severity + Age, data = pdata_avr)

mvals_lmfit_avr <- lmFit(mvals_avr, design = design_avr)
betas_lmfit_avr <- lmFit(betas_avr, design = design_avr)

mvals_ebayes_avr <- eBayes(mvals_lmfit_avr)
betas_ebayes_avr <- eBayes(betas_lmfit_avr)

ACTvREM <- topTable(mvals_ebayes_avr, coef = "SeverityActive", number = "Inf", adjust.method = "BH")
ACTvREM <- cbind(ACTvREM, Beta = coefficients(betas_ebayes_avr)[rownames(ACTvREM), "SeverityActive"], gmset_anno[rownames(ACTvREM), ])

write.csv(ACTvREM, file.path(dmpavrDir, "ACTvREM.csv"))

for(i in 1:10){
  Cairo(width = 600, height = 800, type = "pdf", file = file.path(dmpavrDir, paste0(i, "_", rownames(ACTvREM)[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(ACTvREM)[i], betas = betas_avr, factor_interest = pdata_avr$Severity, enlarged = F, type = "SE", legend = F) + 
          theme(panel.grid.major=element_blank(),
                panel.grid.minor=element_blank()))
  dev.off()
}
```

```{r Heatmap avr}
hm_anno_t50_avr <- data.frame(Status = pdata_avr$Severity, row.names = rownames(pdata_avr))

betas_t50_avr <- betas_avr[rownames(ACTvREM[1:50,]),]
betas_t50_avr_dm <- betas_t50_avr-rowMeans(betas_t50_avr)

pheatmap(betas_t50_avr_dm, annotation_col = hm_anno_t50_avr)
```

```{r Differentially methylated genes avr}
dmgavrDir <- file.path(dmgDir, "avr")
dir.create(dmgavrDir)

ACTvREM_genes <- lapply(strsplit(ACTvREM$UCSC_RefGene_Name, ";"), unique)
names(ACTvREM_genes) <- rownames(ACTvREM)
ACTvREM_genes_nz <- ACTvREM_genes[which(lapply(ACTvREM_genes, length) != 0)]

ACTvREM_genes_nz_df <- data.frame(CpG = gsub("(cg[0-9]{8})[0-9]", "\\1", names(unlist(ACTvREM_genes_nz))), 
                                  Gene = unlist(ACTvREM_genes_nz))
ACTvREM_genes_nz_df$pvalue <- ACTvREM[ACTvREM_genes_nz_df$CpG, "P.Value"]

ACTvREM_genes_nz_list <- split(ACTvREM_genes_nz_df, ACTvREM_genes_nz_df$Gene)

ACTvREM_fisher <- lapply(ACTvREM_genes_nz_list, function(gene){
  aggregation::fisher(gene$pvalue)
})

rm(ACTvREM_genes_nz_list)

ACTvREM_ma <- data.frame(Gene = names(ACTvREM_fisher),
                         Fisher_pval = unlist(ACTvREM_fisher))
ACTvREM_ma$padj <- p.adjust(ACTvREM_ma$Fisher_pval)
ACTvREM_ma <- ACTvREM_ma[order(ACTvREM_ma$Fisher_pval),]
ACTvREM_ma_sig <- ACTvREM_ma[which(ACTvREM_ma$padj<0.05),]

write.csv(ACTvREM_ma, file.path(dmgavrDir, "ACTvREM.csv"))
```

```{r plotting dmgs avr}
ACTvREM_gr <- makeGRangesFromDataFrame(ACTvREM, keep.extra.columns = T, start.field = "pos", end.field = "pos", seqnames.field = "chr")

for(i in 1:length(ACTvREM_ma_sig$Gene)){
  Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(dmgavrDir, paste0(i, "_", ACTvREM_ma_sig$Gene[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[i], tophits_gr = ACTvREM_gr, smooth = T, significance = F))
  dev.off()
}
```

```{r ppi dmgs avr}
dmgavrppiDir <- file.path(dmgavrDir, "ppi")

ACTvREM_ppi <- read.csv(file.path(dmgavrppiDir, "dmg_avr_interactions.tsv"), sep = "\t")

ACTvREM_ppi <- lapply(ACTvREM_ppi, function(interaction){
  interaction %>% 
    dplyr::rename(weight = combined_score,
                  from = node1,
                  to = node2) %>%
    mutate(from = as.character(from),
           to = as.character(to))
})
```


```{r DMR analysis avr}
dmravrDir <- file.path(dmrDir, "avr")
dir.create(dmravrDir)

require(DMRcate)
dmr_anno_ACTvREM <- cpg.annotate(object = as.matrix(mvals_avr),
                                 datatype = "array",
                                 arraytype = "450K",
                                 what = "M",
                                 analysis.type = "differential",
                                 coef = "SeverityActive",
                                 design = design_avr)
dmr_ACTvREM <- dmrcate(dmr_anno_ACTvREM, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_ACTvREM_gr <- extractRanges(dmr_ACTvREM, genome = "hg19")

write.csv(as.data.frame(dmr_ACTvREM_gr), file.path(dmrDir, "dmrs_ACTvREM.csv"))

dmr_plot(dmr_gr = dmr_ACTvREM_gr[1], meth_data = betas_avr, anno_gr = gmset_anno_gr, meth_groups = pdata_avr$Severity, flanks = 0)
```

## Overlap CD vs non-CD and ACT vs REM

```{r dmgs}
require(ggrepel)

dmg_ol <- merge(CDvNon_CD_ma, ACTvREM_ma, by = "Gene")
colnames(dmg_ol) <- c("Gene", "Fisher_pval_cvn", "padj_cvn", "Fisher_pval_avr", "padj_avr")

dmg_ol$Significance <- "NS"
dmg_ol$Significance[dmg_ol$padj_cvn <= 0.05 & dmg_ol$padj_avr <= 0.05] <- "Both"
dmg_ol$Significance[dmg_ol$padj_cvn <= 0.05 & dmg_ol$padj_avr > 0.05] <- "CD vs non-CD"
dmg_ol$Significance[dmg_ol$padj_cvn > 0.05 & dmg_ol$padj_avr <= 0.05] <- "CD-active vs CD-remissive"
dmg_ol$Significance <- factor(dmg_ol$Significance, levels = c("NS", "CD vs non-CD", "CD-active vs CD-remissive", "Both"))
dmg_ol$euclidian <- sqrt((-log10(dmg_ol$Fisher_pval_cvn))^2+(-log10(dmg_ol$Fisher_pval_avr))^2)
dmg_ol <- dmg_ol[order(dmg_ol$euclidian, decreasing = T),]
dmg_ol$label[dmg_ol$Gene %in% c("C6orf25", "NNAT", "BLCAP", "HOXC4")] <- as.character(dmg_ol$Gene[dmg_ol$Gene %in% c("C6orf25", "NNAT", "BLCAP", "HOXC4")])

dmg_ol_plot <- ggplot(dmg_ol, aes(x = -log10(dmg_ol$Fisher_pval_cvn), y = -log10(dmg_ol$Fisher_pval_avr))) +
  geom_point(aes(col = Significance)) +
  geom_text_repel(aes(label = label)) +
  scale_color_manual(values = c(NS = "#D9D5D5", `CD vs non-CD` = "#67d0dd", `CD-active vs CD-remissive` = "#f9665e", Both = "#a885ee")) +
  labs(title = "Comparison CD vs non-CD with CD-active vs CD-remissive",
       subtitle = "-log10(p-value)") +
  xlab("CD vs non-CD") +
  ylab("CD-active vs CD-remissive") +
  theme_bw() +
  theme(panel.grid.major=element_blank(),
        panel.grid.minor=element_blank())
```

```{r overlapping dmgs}
#NNAT and BLCAP
nnat_gr <- makeGRangesFromDataFrame(ACTvREM[grep("NNAT", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
nnat_gr <- nnat_gr[start(nnat_gr) > 36148500 & start(nnat_gr) < 36150000,]

nnat_plot <- dmcomp_plot(gene_of_interest = "NNAT",
                         anno_gr = gmset_anno_gr,
                         tophits1 = CDvNon_CD,
                         tophits2 = ACTvREM,
                         comparison1_name = "CD vs non-CD",
                         comparison2_name = "CD-active vs CD-remissive",
                         meth_data = betas_filtered,
                         meth_groups = pdata_filtered$Severity,
                         ylim = NULL,
                         smooth_m = F)
nnat_enlarged_plot <- dmr_plot(dmr_gr = nnat_gr, 
                               meth_data = betas_filtered, 
                               anno_gr = gmset_anno_gr, 
                               meth_groups = pdata_filtered$Severity, 
                               flanks = 0, 
                               rug = F) +
  scale_color_manual(values=c("#00dae0", "#0000ff", "#ff0000")) +
  theme(axis.title.y = element_blank(),
        legend.pos = "none",
        plot.title = element_blank(),
        plot.subtitle = element_blank())

#MPIG6B (C6orf25)
mpig6b_gr <- makeGRangesFromDataFrame(ACTvREM[grep("C6orf25", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
mpig6b_gr <- mpig6b_gr[start(mpig6b_gr) > 31691000 & start(mpig6b_gr) < 31692000,]

mpig6b_plot <- dmcomp_plot(gene_of_interest = "C6orf25",
                           anno_gr = gmset_anno_gr,
                           tophits1 = CDvNon_CD,
                           tophits2 = ACTvREM,
                           comparison1_name = "CD vs non-CD",
                           comparison2_name = "CD-active vs CD-remissive",
                           meth_data = betas_filtered,
                           meth_groups = pdata_filtered$Severity,
                           ylim = NULL,
                           smooth_m = F)
mpig6b_enlarged_plot <- dmr_plot(dmr_gr = mpig6b_gr, 
                                 meth_data = betas_filtered, 
                                 anno_gr = gmset_anno_gr, 
                                 meth_groups = pdata_filtered$Severity, 
                                 flanks = 0, 
                                 rug = F) +
  scale_color_manual(values=c("#00dae0", "#0000ff", "#ff0000")) +
  theme(axis.title.y = element_blank(),
        legend.pos = "none",
        plot.title = element_blank(),
        plot.subtitle = element_blank())

# HOXC4
hoxc4_gr <- makeGRangesFromDataFrame(ACTvREM[grep("HOXC4", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")
hoxc4_gr <- hoxc4_gr[start(hoxc4_gr) > 54445000 & start(hoxc4_gr) < 54450000,]

hoxc4_plot <- dmcomp_plot(gene_of_interest = "HOXC4",
                          anno_gr = gmset_anno_gr,
                          tophits1 = CDvNon_CD,
                          tophits2 = ACTvREM,
                          comparison1_name = "CD vs non-CD",
                          comparison2_name = "CD-active vs CD-remissive",
                          meth_data = betas_filtered,
                          meth_groups = pdata_filtered$Severity,
                          ylim = NULL,
                          smooth_m = F)
hoxc4_enlarged_plot <- dmr_plot(dmr_gr = hoxc4_gr, 
                                meth_data = betas_filtered, 
                                anno_gr = gmset_anno_gr, 
                                meth_groups = pdata_filtered$Severity, 
                                flanks = 0, 
                                rug = F) +
  scale_color_manual(values=c("#00dae0", "#0000ff", "#ff0000")) +
  theme(axis.title.y = element_blank(),
        legend.pos = "none",
        plot.title = element_blank(),
        plot.subtitle = element_blank())
```

```{r GPR146}
gpr146_gr <- makeGRangesFromDataFrame(ACTvREM[grep("GPR146", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

gpr146_gr[start(gpr146_gr) > 1096000 & start(gpr146_gr) < 1096750,]

gpr146_plot <- dmcomp_plot(gene_of_interest = "GPR146",
                           anno_gr = gmset_anno_gr,
                           tophits1 = CDvNon_CD,
                           tophits2 = ACTvREM,
                           comparison1_name = "CD vs non-CD",
                           comparison2_name = "CD-active vs CD-remissive",
                           meth_data = betas_filtered,
                           meth_groups = pdata_filtered$Severity,
                           ylim = NULL,
                           smooth_m = F)

Cairo(width = 1500, height = 1500, type = "pdf", file = "GPR146_monocytes.pdf", bg = "white", units = "px", dpi = 120)
print(gpr146_plot)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "GPR146_cg20603222_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg20603222", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "GPR146_cg26224354_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg26224354", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()
```

```{r SP140}
sp140_gr <- makeGRangesFromDataFrame(ACTvREM[grep("SP140", ACTvREM$UCSC_RefGene_Name),], keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

sp140_gr <- sp140_gr[start(sp140_gr) >= 231090500,]

sp140_plot <- dmcomp_plot(gene_of_interest = "SP140",
                           anno_gr = sp140_gr,
                           tophits1 = CDvNon_CD,
                           tophits2 = ACTvREM,
                           comparison1_name = "CD vs non-CD",
                           comparison2_name = "CD-active vs CD-remissive",
                           meth_data = betas_filtered,
                           meth_groups = pdata_filtered$Severity,
                           ylim = NULL,
                           smooth_m = F)

Cairo(width = 1500, height = 1500, type = "pdf", file = "SP140_monocytes.pdf", bg = "white", units = "px", dpi = 120)
print(sp140_plot)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg03887528_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg03887528", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg05564251_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg05564251", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()

Cairo(width = 600, height = 600, type = "pdf", file = "SP140_cg04579254_monocytes.pdf", bg = "white", units = "px", dpi = 120)
cpg_strip_plot(cpg_num = "cg04579254", betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE", legend = F)
dev.off()
```

# Public data

```{r setup public data}
publicDir <- file.path(outputDir, "public")
dir.create(publicDir)
```

## Methyl-Seq: GSE73788

```{r setup methylseq}
mseqDir <- file.path(publicDir, "methylseq")
dir.create(mseqDir)
```

```{r gse73788 import and processing}
gse73788 <- read.csv(file.path("data", "methylseq", "GSE73788_Mo1_Mo2_Mo3_merged.csv"), stringsAsFactors = F)
gse73788$seqnames <- gsub("(chr[0-9XYM]+):[0-9]+-[0-9]+", "\\1", gse73788$Position)
gse73788$start <- as.numeric(gsub("chr[0-9XYM]+:([0-9]+)-[0-9]+", "\\1", gse73788$Position))
gse73788$end <- as.numeric(gsub("chr[0-9XYM]+:[0-9]+-([0-9]+)", "\\1", gse73788$Position))
gse73788_gr <- makeGRangesFromDataFrame(gse73788, keep.extra.columns = T)

gse73788_meth <- data.frame(gse73788[, c("Mo1", "Mo2", "Mo3")])
gse73788_meth_l2i <- -log2(gse73788_meth+1)
gse73788_groups <- c("Classical", "Intermediate", "Non-classical")
```

```{r visualize}
dmgs_sig <- union(ACTvREM_ma_sig$Gene, CDvNon_CD_ma_sig$Gene)

for(i in 41:length(dmgs_sig)){
  goi_gr <- range(gmset_anno_gr[grep(paste0("(^|;)", dmgs_sig[i], "($|;)"), gmset_anno_gr$UCSC_RefGene_Name),], ignore.strand = T)
  
  Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(mseqDir, paste0(dmgs_sig[i], ".pdf")), bg = "white", units = "px", dpi = 120)
  print(dmrma_plot(roi_gr = goi_gr,
                   ownmeth_data = betas_filtered,
                   ownmeth_anno_gr = gmset_anno_gr,
                   ownmeth_groups = pdata_filtered$Severity,
                   pubmeth_data = gse73788_meth_l2i,
                   pubmeth_anno_gr = gse73788_gr,
                   pubmeth_groups = gse73788_groups))
  dev.off()
}
```

# Primers DMR

```{r primer}
fig1propDir <- file.path(proposalDir, "fig1")
dir.create(fig1propDir)

tab1propDir <- file.path(proposalDir, "table1")
dir.create(tab1propDir)


require(GenomicRanges)
amplicons <- data.frame(chr = rep("chr7", 9), 
                        start = c(start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 247-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 744-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 733-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 463-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 744-1,
                                  start(dmr_CDvNon_CD_gr[1]) + 246-1),
                        end = c(start(dmr_CDvNon_CD_gr[1]) + 479,
                                start(dmr_CDvNon_CD_gr[1]) + 475,
                                start(dmr_CDvNon_CD_gr[1]) + 476,
                                start(dmr_CDvNon_CD_gr[1]) + 471,
                                start(dmr_CDvNon_CD_gr[1]) + 976,
                                start(dmr_CDvNon_CD_gr[1]) + 976,
                                start(dmr_CDvNon_CD_gr[1]) + 611,
                                start(dmr_CDvNon_CD_gr[1]) + 977,
                                start(dmr_CDvNon_CD_gr[1]) + 475),
                        L_tm = c(54.95, 54.95, 54.95, 54.95, 54.71, 52.14, 57.38, 54.71, 56.22),
                        R_tm = c(58.17 , 59.30, 58.99, 59.92, 54.90, 54.90, 57.88, 56.23, 59.30),
                        L_GC = c(40.00, 40.00, 40.00, 40.00, 53.57, 40.00, 60.00, 53.57, 38.46),
                        R_GC = c(60.00, 56.00, 56.00, 57.69, 58.33, 58.33, 53.85, 56.00, 56.00),
                        L_seq = c("TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "TTAGAGAATTTTTGGATTTGTAAGT", "GAGGAGTTGTTATTTTTATTTTATTTAA", "GTAGAGTGTTTTAGAAATTTTTTTT", "GGGAGTTGTGTATATTTAGGTTGAG", "GAGGAGTTGTTATTTTTATTTTATTTAA", "TTTAGAGAATTTTTGGATTTGTAAGT"),
                        R_seq = c("AAACTATACCAAAAAAACCCTCAAC", "TATACCAAAAAAACCCTCAACCTAA", "CTATACCAAAAAAACCCTCAACCTA", "CCAAAAAAACCCTCAACCTAAATATA", "AAAACTATATCTCCCTTACCACAC", "AAAACTATATCTCCCTTACCACAC", "CTATAAAAAACAATACAACCCTCTCC", "AAAAACTATATCTCCCTTACCACAC", "TATACCAAAAAAACCCTCAACCTAA"),
                        tm = c(64.7, 64.8, 64.8, 64.5, 64.7, 64.6, 64.6, 64.7, 64.8),
                        prod_nCpGs = c(11, 11, 11, 11, 11, 10, 6, 11, 11),
                        ymin = seq(0, 4, by = 0.5),
                        ymax = seq(0.5, 4.5, by = 0.5))

amplicons$name <- paste0("amp", 1:9)
amplicons$prod_length <- amplicons$end - amplicons$start
amplicons$L_bps <- nchar(as.character(amplicons$L_seq))
amplicons$R_bps <- nchar(as.character(amplicons$R_seq))

amplicons <- amplicons[,c("chr", "start", "end", "name", "L_tm", "R_tm", "tm", "L_GC", "R_GC", "L_bps", "R_bps", "prod_nCpGs", "prod_length", "L_seq", "R_seq", "ymin", "ymax")]
write.csv(amplicons, file.path(tab1propDir, "table1.csv"))

dmr1 <- dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Status1, flanks = 0) + theme(legend.pos = "top", axis.title.x=element_blank(), axis.text.x=element_blank(), axis.ticks.x=element_blank())
amps <- ggplot(amplicons, aes(xmin = start, xmax = end, ymin = ymin, ymax = ymax)) + 
  geom_rect(alpha = 0.25, col = "black") + 
  geom_text(aes(x=start+(end-start)/2, y=ymin+(ymax-ymin)/2, label=name), size = 4) +
  #facet_wrap(~name, ncol = 1, strip.position = "left") +
  theme_bw() +
  xlim(start(dmr_CDvNon_CD_gr[1]), end(dmr_CDvNon_CD_gr[1])) +
  theme(axis.title.y = element_blank(),
        axis.title.x = element_blank(),
        axis.text.y = element_blank(),
        axis.ticks.y = element_blank(),
        panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        axis.text = element_text(size = 12), 
        axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 750, height = 1000, type = "png", file = file.path(fig1propDir, "fig1.png"), bg = "white", units = "px", dpi = 120)
print(ggarrange(dmr1, amps, ncol = 1, nrow = 2, align = "v"))
dev.off()

Cairo(width = 1200, height = 1600, type = "pdf", file = file.path(fig1propDir, "fig1.pdf"), bg = "white", units = "px", dpi = 120)
print(ggarrange(dmr1, amps, ncol = 1, nrow = 2, align = "v"))
dev.off()
```

```{r MethPrimer}
library(RCurl)
library(htmltidy)

methprimer <- readHTMLTable("/home/ayliyim/archive/projects/PRJ0000002_CDMON/data/primers/methprimer_dmr.html")
```

#Figures

```{r Figure 1}
require(ggpubr)
require(Cairo)

fig1Dir <- file.path(manuscriptDir, "fig1")
dir.create(fig1Dir)

#A
Cairo(width = 1250, height = 1500, type = "pdf", file = file.path(fig1Dir, "fig1a.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap(betas_t50_cvn_dm, annotation_col = hm_anno_t50_cvn, show_colnames = F)
dev.off()

#B
CDvNon_CD_ma_t100 <- CDvNon_CD_ma[1:100,]
CDvNon_CD_ma_t100$Gene <- factor(CDvNon_CD_ma_t100$Gene, levels = CDvNon_CD_ma_t100$Gene)
CDvNon_CD_ma_t100$Significance <- CDvNon_CD_ma_t100$padj<0.05

Cairo(width = 3000, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1b.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(CDvNon_CD_ma_t100, aes(x = Gene, y = -log10(Fisher_pval), fill = Significance)) +
  geom_bar(stat = "identity") +
  ylab("-log10(p-value)") +
  theme_bw() +
  theme(panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1),
        axis.title.x = element_blank())
dev.off()

#C
Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1c1.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[1], tophits_gr = CDvNon_CD_gr, smooth = T, significance = F))
dev.off()

wdfy4 <- CDvNon_CD_gr[grep("WDFY4", CDvNon_CD_gr$UCSC_RefGene_Name),]
wdfy4 <- wdfy4[start(wdfy4) < 49920000,]

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1c2.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[2], tophits_gr = wdfy4, smooth = T, significance = F))
dev.off()

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1c3.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[3], tophits_gr = CDvNon_CD_gr, smooth = T, significance = F))
dev.off()

gstt1 <- CDvNon_CD_gr[grep("GSTT1", CDvNon_CD_gr$UCSC_RefGene_Name),]
gstt1 <- gstt1[start(gstt1) > 24382000,]

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1c4.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[4], tophits_gr = gstt1, smooth = F, significance = F))
dev.off()

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig1Dir, "fig1c5.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = CDvNon_CD_ma_sig$Gene[5], tophits_gr = CDvNon_CD_gr, smooth = T, significance = F))
dev.off()
```

```{r Figure 2}
fig2Dir <- file.path(manuscriptDir, "fig2")
dir.create(fig2Dir)

#A
Cairo(width = 1250, height = 1500, type = "pdf", file = file.path(fig2Dir, "fig2a.pdf"), bg = "white", units = "px", dpi = 120)
pheatmap(betas_t50_avr_dm, annotation_col = hm_anno_t50_avr, show_colnames = F)
dev.off()

#B
ACTvREM_ma_t100 <- ACTvREM_ma[1:100,]
ACTvREM_ma_t100$Gene <- factor(ACTvREM_ma_t100$Gene, levels = ACTvREM_ma_t100$Gene)
ACTvREM_ma_t100$Significance <- ACTvREM_ma_t100$padj<0.05

Cairo(width = 3000, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2b.pdf"), bg = "white", units = "px", dpi = 120)
ggplot(ACTvREM_ma_t100, aes(x = Gene, y = -log10(Fisher_pval), fill = Significance)) +
  geom_bar(stat = "identity") +
  ylab("-log10(p-value)") +
  theme_bw() +
  theme(panel.grid.major=element_blank(),
        panel.grid.minor=element_blank(),
        axis.text.x = element_text(angle = 90, hjust = 1),
        axis.title.x = element_blank())
dev.off()

#C
Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2c1.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[1], tophits_gr = ACTvREM_gr, smooth = T, significance = F))
dev.off()

serpinf1 <- ACTvREM_gr[grep("SERPINF1", ACTvREM_gr$UCSC_RefGene_Name),]
serpinf1 <- serpinf1[start(serpinf1) < 1666000 & start(serpinf1) > 1664400,]

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2c2.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[3], tophits_gr = serpinf1, smooth = T, significance = F))
dev.off()

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2c3.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[4], tophits_gr = ACTvREM_gr, smooth = T, significance = F))
dev.off()

Cairo(width = 1500, height = 750, type = "pdf", file = file.path(fig2Dir, "fig2c4.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_plot(gene_of_interest = ACTvREM_ma_sig$Gene[5], tophits_gr = ACTvREM_gr, smooth = T, significance = F))
dev.off()
```


```{r Figure 3}
fig3Dir <- file.path(manuscriptDir, "fig3")
dir.create(fig3Dir)

#A
Cairo(width = 1500, height = 1000, type = "pdf", file = file.path(fig3Dir, "fig3a.pdf"), bg = "white", units = "px", dpi = 120)
print(dmg_ol_plot)
dev.off()

#B
Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(fig3Dir, "fig3b1.pdf"), bg = "white", units = "px", dpi = 120)
print(mpig6b_plot)
dev.off()

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(fig3Dir, "fig3b2.pdf"), bg = "white", units = "px", dpi = 120)
print(mpig6b_enlarged_plot)
dev.off()

#C
Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(fig3Dir, "fig3c1.pdf"), bg = "white", units = "px", dpi = 120)
print(nnat_plot)
dev.off()

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(fig3Dir, "fig3c2.pdf"), bg = "white", units = "px", dpi = 120)
print(nnat_enlarged_plot)
dev.off()

#D
Cairo(width = 1500, height = 1500, type = "pdf", file = file.path(fig3Dir, "fig3d1.pdf"), bg = "white", units = "px", dpi = 120)
print(hoxc4_plot)
dev.off()

Cairo(width = 1500, height = 500, type = "pdf", file = file.path(fig3Dir, "fig3d2.pdf"), bg = "white", units = "px", dpi = 120)
print(hoxc4_enlarged_plot)
dev.off()


```

```{r Figure S1}
figS1Dir <- file.path(manuscriptDir, "figS1")
dir.create(figS1Dir)

Cairo(width = 1000, height = 750, type = "pdf", file = file.path(figS1Dir, "figS1.pdf"), bg = "white", units = "px", dpi = 120)
print(mvals_filtered_svd_plotobj)
dev.off()
```

