---
title: "Differential expression analyses public data"
author: "Andrew Y.F. Li Yim"
date: "January 17, 2019"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

In the previous analyses, we performed differential methylation analysis where we compared CD14+ monocytes obtained from CD with non-CD as well as with CD-active with CD-remissive where we identified differentially methylated genes (DMGs) for both comparisons. One snag we ran into was the fact that CD14+ monocytes are not a homogeneous population and in fact consist of three subpopulations, namely the classical, intermediate and non-classical. Moreover, from the literature, we know that the proportions of these monocyte populations are associated with CD and CD-activity [Koch et al. 2010](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2909416/). In our methylation analyses we downloaded the data from [Zawada et al. 2016](https://www.ncbi.nlm.nih.gov/pubmed/27018948), but found the results to be inconclusive. To say something sensible about our DMGs, we could make an abstraction by looking at gene expression. For this to make sense, we must make the assumption that gene expression and DNA methylation are inversely correlated (while this is typically the case, this is not true for all). Luckily, such expression data on monocyte subpopulations exist with the often-used dataset by [Monaco et al. 2019](https://www.ncbi.nlm.nih.gov/pubmed/30726743) being an example thereof. I therefore downloaded the their raw RNA sequencing data from [GSE107011](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE107011) and aligned it using scripts found in the same directory. After having obtained counts, I started to do the analysis. 

Essentially, we are interested in whether the DMGs are associated with changes in monocyte populations. To test whether a gene is associated with monocyte populations, we can utilize the loglikelihood ratio test, where we define the full model as `covariates + monocyte` and the reduced model as `covariates`.

```{r packages}
require(Cairo)
require(ggplot2)
require(plotly)
require(ggpubr)
require(NDlib) # Can be found on my Github
```

```{r setup}
deaDir <- file.path("output", "02_DE_analyses")
dir.create(deaDir)

rdsDir <- file.path(deaDir, "rds")
dir.create(rdsDir)
```

```{r colors}
#Function sourced from https://stackoverflow.com/questions/8197559/emulate-ggplot2-default-color-palette
gg_color_hue <- function(n) {
  hues = seq(15, 375, length = n + 1)
  hcl(h = hues, l = 65, c = 100)[1:n]
}

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
```

```{r samples}
samples <- read.csv(file.path("data", "samples", "samplesheet_gse107011_PRJ0000002_CDMON_v1.csv"), stringsAsFactors = F)
rownames(samples) <- samples$SRR
samples$Monocytes <- factor(samples$Monocytes, levels = c("Classical", "Intermediate", "Non-classical"))
```

# Import and preprocessing

Previously, we performed read counting using the `featureCount` tool found in the `Subread` package. The `featureCount` tool adds some extra columns, which are not easily parsable, we will therefore ignore these columns.

```{r count processing}
require(DESeq2)
require(biomaRt)

#Annotations
enshsbm <- useMart("ensembl", dataset = "hsapiens_gene_ensembl")
ens2symbol <- getBM(attributes = c("ensembl_gene_id", "entrezgene_id", "hgnc_symbol"), mart = enshsbm)

#Counts
gcounts_raw <- read.csv(file.path("output", "02_DE_analyses", "counts", "GRCh37", "counts.txt"), sep = "\t", skip = 1)
gcounts <- gcounts_raw[,c(7:ncol(gcounts_raw))]
rownames(gcounts) <- gcounts_raw$Geneid

#Remove genes with 0 counts
gcounts <- gcounts[rowSums(gcounts) != 0,]

colnames(gcounts) <- gsub("^.+\\.(SRR[0-9]+)\\.bam", "\\1", colnames(gcounts))
gcounts <- gcounts[,rownames(samples)]
```

Having performed some basic cleaning on the data, we will prepare the DESeqDataSet for differential expression analysis.

```{r prepare dds}
#Remove genes with low read counts
dds <- DESeqDataSetFromMatrix(countData = gcounts,
                              colData = samples,
                              design = ~ Monocytes + Source)

dds <- DESeq(dds, test = "LRT", reduced = ~Source)

#Useful for visualization purposes
rld <- rlog(dds)
rld_counts <- assay(rld)
```

# Exploratory data analyses

While not necessary, we can still perform some degree of exploratory analyses to investigate general structures.

```{r sample pairwise analysis}
require(pheatmap)
edaDir <- file.path(deaDir, "eda")
dir.create(edaDir)

rld_cor <- cor(rld_counts)

rld_hm_anno <- data.frame(Monocyte = colData(rld)$Monocytes,
                          Origin = colData(rld)$Source, 
                          row.names = colnames(rld))

Cairo(file = file.path(edaDir, "pairwise_sample_correlations.pdf"), type = "pdf", units = "px", width = 1200, height = 1000, dpi = 90, bg = "white")
pheatmap(rld_cor, annotation_row = rld_hm_anno, annotation_col = rld_hm_anno)
dev.off()
```

```{r principal component analysis}
rld_dm <- rld_counts-rowMeans(rld_counts)

rld_svd <- svd(t(rld_dm))

#Percentage variance
rld_svd_pct_var <- round(rld_svd$d/sum(rld_svd$d)*100, 2)

#Plot
rld_svd_df <- data.frame(PC1 = rld_svd$u[,1],
                         PC2 = rld_svd$u[,2],
                         Monocyte = colData(rld)$Monocytes,
                         Origin = colData(rld)$Source,
                         Sample = colData(rld)$SRR)

pca_plotobj <- ggplot(rld_svd_df, aes(x = PC1, y = PC2, col = Monocyte, shape = Origin, label = Sample)) +
  geom_point(size = 3) +
  theme_bw() +
  xlab(paste0("PC1 (", rld_svd_pct_var[1], "%)")) +
  ylab(paste0("PC2 (", rld_svd_pct_var[2], "%)")) +
  theme(legend.pos = "bottom",
        panel.grid.major = element_blank(),
        panel.grid.minor = element_blank())

Cairo(file = file.path(edaDir, "pca.pdf"), type = "pdf", units = "px", width = 1000, height = 1000, dpi = 90, bg = "white")
print(pca_plotobj)
dev.off()

ggplotly(pca_plotobj)
```

```{r sex estimation}
transcript_strip_plot(id = "ENSG00000229807", counts = rld_counts, factor_interest = colData(rld)$Source)

rld_counts["ENSG00000229807",]
```

There appear to be 2 males and 2 females.

# Differential expression analyses

```{r differential analyses setup}
degDir <- file.path(deaDir, "deg")
dir.create(degDir)
```

```{r subset associated genes}
lrtDir <- file.path(degDir, "lrt")
dir.create(lrtDir)

degs_lrt <- as.data.frame(results(dds))
degs_lrt <- degs_lrt[!is.na(degs_lrt$padj),]
degs_lrt$ENS <- rownames(degs_lrt)
degs_lrt <- merge(degs_lrt, unique(ens2symbol[,c(1, 3)]), by.x = "ENS", by.y = "ensembl_gene_id", all.x = T, all.y = F)
degs_lrt <- degs_lrt[order(degs_lrt$pvalue),]
head(degs_lrt)
write.csv(degs_lrt, file.path(lrtDir, "degs_lrt.csv"))

lrtsigDir <- file.path(lrtDir, "sig_plots")
dir.create(lrtsigDir)

print(transcript_strip_plot(id = degs_lrt[i,"ENS"],
                            counts = assay(rld),
                            factor_interest = colData(rld)$Monocytes,
                            y_lab = "log2(Counts)",
                            title = degs_lrt[i,"mgi_symbol"]))
```

Next we investigated whether the expression of the DMGs are associated with monocyte subsets. 

```{r dmgs}
dmgdegDir <- file.path(degDir, "dmgdeg", "200225")
dir.create(dmgdegDir)

dmgdeg_cvnDir <- file.path(dmgdegDir, "cvn")
dir.create(dmgdeg_cvnDir)

#CD vs non-CD
cvn_dmgs_sig_anno <- merge(data.frame(gene = CDvNon_CD_ma_sig$Gene), ens2symbol, by.x = "gene", by.y = "hgnc_symbol")

cvn_degs_lrt_ol <- degs_lrt[which(degs_lrt$ENS %in% cvn_dmgs_sig_anno$ensembl_gene_id),]

cvn_plotlist <- list()

for(i in 1:nrow(cvn_degs_lrt_ol)){
  cvn_plotlist[[i]] <- transcript_strip_plot(id = cvn_degs_lrt_ol$ENS[i],
                                             counts = assay(rld),
                                             factor_interest = colData(rld)$Monocytes,
                                             title = cvn_degs_lrt_ol$hgnc_symbol[i], 
                                             type = "SE") +
    labs(subtitle = paste0("p-value = ", formatC(cvn_degs_lrt_ol$pvalue[i], format = "e", digits = 2))) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.title.y = element_blank(),
          axis.ticks.x = element_blank())
}

cvn_plot_objs <- ggarrange(plotlist = cvn_plotlist, nrow = 2, ncol = 5, align = "hv", common.legend = T, legend = "none")

#CD-active vs CD-remissive
avr_dmgs_sig_anno <- merge(data.frame(gene = ACTvREM_ma_sig$Gene), ens2symbol, by.x = "gene", by.y = "hgnc_symbol")

avr_degs_lrt_ol <- degs_lrt[which(degs_lrt$ENS %in% avr_dmgs_sig_anno$ensembl_gene_id),]

avr_plotlist <- list()

for(i in 1:nrow(avr_degs_lrt_ol)){
  avr_plotlist[[i]] <- transcript_strip_plot(id = avr_degs_lrt_ol$ENS[i],
                                             counts = assay(rld),
                                             factor_interest = colData(rld)$Monocytes,
                                             title = avr_degs_lrt_ol$hgnc_symbol[i], 
                                             type = "SE") +
    labs(subtitle = paste0("p-value = ", formatC(avr_degs_lrt_ol$pvalue[i], format = "e", digits = 2))) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          axis.title.y = element_blank(),
          axis.ticks.x = element_blank())
}

avr_plot_objs <- ggarrange(plotlist = avr_plotlist, nrow = 2, ncol = 5, align = "hv", common.legend = T, legend = "none")
```

For multiple of the DMGs, their expression is associated with monocyte subsets. It is enticing to suggest therefore that the methylation might also be affected by differences in populations. Turning that argument around, the observed differences in methylation for the subset-associated DMGs might simply be the differences in monocyte populations. Interestingly, the non-significant DMGs, might then be unassociated with monocyte populations, which would warrant further research. Either way, substantial validation experiments would be necessary to corroborate our observations.

# Figures

```{r fig4 v1}
rld_counts_dm <- rld_counts-rowMeans(rld_counts)

rld_counts_dm <- rld_counts_dm[,order(colData(rld)$Monocytes)]
rld_counts_dm_dmg <- rld_counts_dm[rownames(rld_counts_dm) %in% degs_lrt_ol$ENS,]

degs_lrt_ol <- degs_lrt_ol[match(rownames(rld_counts_dm_dmg), degs_lrt_ol$ENS),]

pheatmap_cols <- list(Monocyte = c(Classical = gg_color_hue(3)[1],
                                   Intermediate = gg_color_hue(3)[2],
                                   `Non-classical` = gg_color_hue(3)[3]))

degs_lrt_anno <- data.frame(pvalue = -log10(degs_lrt_ol$pvalue), significant = (degs_lrt_ol$pvalue<0.05)*1, row.names = degs_lrt_ol$ENS)

Cairo(file = file.path(dmgdegDir, "dmg_de.pdf"), type = "pdf", units = "px", width = 800, height = 600, dpi = 90, bg = "white")
pheatmap(rld_counts_dm_dmg, 
         annotation_col = rld_hm_anno, 
         annotation_row = degs_lrt_anno,
         cluster_cols = F, 
         labels_row = degs_lrt_ol$hgnc_symbol, 
         labels_col = colData(rld)[order(colData(rld)$Monocytes), "GSM"],
         annotation_colors = pheatmap_cols)
dev.off()

```

```{r fig4 v2}
fig4v2Dir <- file.path("docs", "manuscript", "data", "200128", "fig4", "v2")
dir.create(fig4v2Dir)

fig4v2 <- ggarrange(cvn_plot_objs, avr_plot_objs, nrow = 2, ncol = 1, labels = c("a)", "b)"))

Cairo(file = file.path(fig4v2Dir, "fig4v2.pdf"), type = "pdf", units = "px", width = 1200, height = 1200, dpi = 90, bg = "white")
print(fig4v2)
dev.off()
```