---
title: "DM analysis PRJ0000002_CDMON"
author: "Andrew Li Yim"
date: "November 4, 2015"
output: html_document
---

In this report we will perform the differential methylation analyses on CD14+ monocytes derived from CD and non-CD patients. The CD patients can be further subdivided into active and remissive CD.

```{r setup, echo = F}
require(minfi)
samplesDir <- file.path("data", "samples")
commondataDir <- file.path("/home", "ayliyim", "archive", "common_data")
outputDir <- file.path("output", "01_DM_analyses")
dir.create(outputDir)

manuscriptDir <- file.path("publications", "manuscript")
dir.create(manuscriptDir)

samplesheet <- read.metharray.sheet(base = samplesDir, pattern = "samplesheet_PRJ0000002_CDMON_V5.csv")
samplesheet$Basename <- samplesheet$Path
samplesheet$Sex <- "F"

rgset <- read.metharray.exp(targets = samplesheet)
```

```{r sample information}
require(dplyr)
samplesheet %>% 
  select(Age, Severity) %>%
  group_by(Severity) %>%
  summarize(mu = mean(Age), stdev = (sd(Age)))
```

```{r Quality control, echo = FALSE}
edaDir <- file.path(outputDir, "eda")
dir.create(edaDir)

require(MethylAid)
methylaid_summarized <- summarize(samplesheet)
visualize(methylaid_summarized)

require(shinyMethyl)
shinymethyl_summarized <- shinySummarize(rgset)
runShinyMethyl(shinymethyl_summarized)

require(ewastools)
ewastools_samples <- read_idats(samplesheet$Basename)
ewastools_betas <- dont_normalize(ewastools_samples)
ewastools_snps <- call_genotypes(ewastools_betas[ewastools_samples$manifest[probe_type == 'rs', index], ], learn = F)
check_snp_agreement(genotypes = ewastools_snps$gamma, weights = 1 - ewastools_snps$outliers, donor_ids = samplesheet$Donor_ID, sample_ids = samplesheet$Sample_ID)
```

According to MethylAid, all samples pass the quality control. Additionally, the ewastools package was used to confirm that the samples were obtained from the same origin.
As the sample pertains peripheral blood, we can utilize the estimateCellCounts function from minfi to see whether we observe any differences in estimated cell populations between the various groups.

```{r Blood cell distribution}
require(FlowSorted.Blood.450k)
require(ggplot2)
require(Cairo)

cellCounts <- estimateCellCounts(rgSet = rgset, 
                                 compositeCellType = "Blood",
                                 processMethod = "auto", 
                                 probeSelect = "auto")
cellCounts <- data.frame(cellCounts, severity = pData(rgset)$Severity, array = rownames(cellCounts))

cellCounts_melt <- reshape2::melt(cellCounts, id.vars = c("severity", "array"), id.vals = c("Response"))

cellCounts_plot <- ggplot(cellCounts_melt, aes(x = severity, y = value)) +
  geom_boxplot(outlier.shape = NA) +
  geom_jitter() +
  facet_wrap(~variable) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

Cairo(width = 1000, height = 1000, type = "png", file = file.path(edaDir, "estimatedCellCounts.png"), bg = "white", units = "px", dpi = 120)
print(cellCounts_plot)
dev.off()
```

The cell distribution accurately guesses that the samples pertain monocytes. There appears to be one sample however that appears more CD4T-like.

```{r CD4T samples}
require(dplyr)

CD4T_array_ID <- cellCounts_melt %>% 
  filter(variable == "Mono" & value < 0.25) %>% 
  select(array)

pData(rgset)[CD4T_array_ID$array, "Sample_ID"]
```

We will next investigate whether this particular sample is an outlier in the PCA as well. Before that, we will remove the regular outliers.

```{r Preprocessing}
gmset <- preprocessFunnorm(rgSet = rgset)
gmset_nosnpcpg <- gmset[with(getSnpInfo(gmset), which(CpG_maf < 0.01 | is.na(CpG_maf))),]
gmset_nosnpsbe <- gmset_nosnpcpg[with(getSnpInfo(gmset_nosnpcpg), which(SBE_maf < 0.01 | is.na(SBE_maf))),]
gmset_nosex <- gmset_nosnpsbe[-which(getAnnotation(gmset_nosnpsbe)$chr %in% c("chrX", "chrY")),]

pmprobes <- read.csv(file.path(commondataDir, "IlluminaHumanMethylationBeadChip450k", "Non-specific-probes-Illumina450k.csv"), stringsAsFactors = F, header = T)[,1]
gmset_filtered <- gmset_nosex[!names(gmset_nosex) %in% pmprobes,]

mvals <- getM(gmset_filtered)

gmset_filtered <- gmset_filtered[-which(abs(mvals) == Inf, arr.ind = T)[,1],]

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

Are there any sex discrepancies?

```{r Annotated and predicted sex}
require(ggplot2)
require(dplyr)
require(ggrepel)

pData(gmset_filtered) %>% 
  data.frame %>% 
  select(Sample_ID, Sex, predictedSex, xMed, yMed) %>%
  mutate(Sample_label = ifelse(Sex == predictedSex, "", Sample_ID)) %>% 
  ggplot(aes(x = xMed, y = yMed, col = Sex, label = Sample_label)) +
  geom_text_repel(show.legend = F) +
  geom_point(size = 4) +
  theme_bw() +
  coord_fixed() +
  ggtitle("Sex chromosome probe median intensities") +
  xlab("chrX") +
  ylab("chrY") +
  theme(text = element_text(size=20))
```

```{r PCA}
require(plotly)

mvals_dm <- mvals - rowMeans(mvals)
mvals_svd <- svd(t(mvals_dm))

mvals_svd_df <- data.frame(Sample_ID = pData(gmset_filtered)$Sample_ID,
                           Origin = pData(gmset_filtered)$Donor_ID,
                           Age = pData(gmset_filtered)$Age,
                           Severity = pData(gmset_filtered)$Severity,
                           Sex = pData(gmset_filtered)$predictedSex,
                           PC1 = mvals_svd$u[,1],
                           PC2 = mvals_svd$u[,2])

mvals_svd_plotobj <- ggplot(mvals_svd_df, aes(x = PC1, y = PC2, Label = Sample_ID, Origin = Origin, Severity = Severity, Age = Age, Sex = Sex)) +
  geom_point(aes(col = Origin)) +
  theme_bw() +
  theme(text = element_text(size=20))

ggplotly(mvals_svd_plotobj)
```

We again observe an outlier on PC1.

```{r PCA outlier}
mvals_svd_df[which.max(mvals_svd_df$PC1),]
```

This outlier appears to be HC1 again. Due to its outlier status, we will therefore remove HC1 from further analyses.

```{r HC1 removal}
gmset_filtered <- gmset_filtered[, pData(gmset_filtered)$Sample_ID != "HC1"]
pData(gmset_filtered)$Severity <- factor(pData(gmset_filtered)$Severity, levels = c("Non_CD", "Remissive", "Active"))

betas <- getBeta(gmset_filtered)
mvals <- getM(gmset_filtered)
```

There are two samples derived from the same donor: CDACT-1a and CDACT-1b. As there is no clear way of choosing one over the other, we will take the mean.

```{r CDACT1 aggregation}
mvals_filtered <- data.frame(CDACT1 = rowMeans(mvals[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), mvals[, 3:ncol(mvals)])
betas_filtered <- data.frame(CDACT1 = rowMeans(betas[, which(pData(gmset_filtered)$Donor_ID %in% "CDACT-1", arr.ind = T)]), betas[, 3:ncol(betas)])

pdata_filtered <- pData(gmset_filtered)[-1,]
colnames(mvals_filtered) <- colnames(betas_filtered) <- rownames(pdata_filtered)
```

```{r DM analyses preparation}
require(limma)

design_mat <- model.matrix(~0 + Severity + Age, data = pdata_filtered)
colnames(design_mat) <- c("Non_CD", "Remissive", "Active", "Age")

mvals_lmfit <- lmFit(mvals_filtered, design = design_mat)
betas_lmfit <- lmFit(betas_filtered, design = design_mat)

contrast_mat <- makeContrasts(
  CDvNon_CD = (Remissive + Active)/2 - Non_CD,
  ACTvREM = Active - Remissive,
  levels = design_mat
)

mvals_contrastfit <- contrasts.fit(mvals_lmfit, contrast_mat)
mvals_ebayes <- eBayes(mvals_contrastfit)

betas_contrastfit <- contrasts.fit(betas_lmfit, contrast_mat)
betas_ebayes <- eBayes(betas_contrastfit)
```

```{r DMP analysis}
dmpDir <- file.path(outputDir, "dmps")
dir.create(dmpDir)

require(NDlib)

gmset_anno <- getAnnotation(gmset_filtered)
gmset_anno_gr <- makeGRangesFromDataFrame(gmset_anno, keep.extra.columns = T, seqnames.field = "chr", start.field = "pos", end.field = "pos")

CDvNon_CD <- topTable(mvals_ebayes, coef = "CDvNon_CD", number = "Inf", adjust.method = "BH")
CDvNon_CD <- cbind(CDvNon_CD, Beta = coefficients(betas_ebayes)[rownames(CDvNon_CD), "CDvNon_CD"], gmset_anno[rownames(CDvNon_CD), ])
ACTvREM <- topTable(mvals_ebayes, coef = "ACTvREM", number = "Inf", adjust.method = "BH")
ACTvREM <- cbind(ACTvREM, Beta = coefficients(betas_ebayes)[rownames(ACTvREM), "ACTvREM"], gmset_anno[rownames(ACTvREM), ])

write.csv(CDvNon_CD, file.path(dmpDir, "CDvNon_CD.csv"))
write.csv(ACTvREM, file.path(dmpDir, "ACTvREM.csv"))

for(i in 1:10){
  Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(dmpDir, paste0(i, "_", rownames(CDvNon_CD)[i], "_CDvNon_CD.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(CDvNon_CD)[i], betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F))
  dev.off()
  
  Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(dmpDir, paste0(i, "_", rownames(ACTvREM)[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(ACTvREM)[i], betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F))
  dev.off()
}
```

##CD vs non-CD
```{r Surrounding CpGs cg17206772}
cg17206772_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[1])
cg17206772_area_gr <- gmset_anno_gr[(cg17206772_area-5):(cg17206772_area+5),]

dmr_plot(dmr_gr = range(cg17206772_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)
```

```{r Surrounding CpGs cg16921727}
cg16921727_area <- which(rownames(gmset_anno) == rownames(CDvNon_CD)[2])
cg16921727_area_gr <- gmset_anno_gr[(cg16921727_area-5):(cg16921727_area+5),]

dmr_plot(dmr_gr = range(cg16921727_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)
```

##ACT vs REM

```{r Surrounding CpGs cg06521852}
cg06521852_area <- which(rownames(gmset_anno) == rownames(ACTvREM)[2])
cg06521852_area_gr <- gmset_anno_gr[(cg06521852_area-5):(cg06521852_area+5),]

dmr_plot(dmr_gr = range(cg06521852_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity, flanks = 1)
```

```{r Overrepresentation analyses}
pathwayDir <- file.path(outputDir, "pathways")
dir.create(pathwayDir)

require(missMethyl)

top100_CDvNon_CD_kegg <- gometh(sig.cpg = rownames(CDvNon_CD)[1:100], all.cpg = rownames(gmset_anno), collection = "KEGG", array.type = "450K")
top100_CDvNon_CD_kegg <- top100_CDvNon_CD_kegg[order(top100_CDvNon_CD_kegg$P.DE),]
write.csv(top100_CDvNon_CD_kegg, file.path(pathwayDir, "CDvNon_CD_top100_kegg.csv"))

top100_ACTvREM_kegg <- gometh(sig.cpg = rownames(ACTvREM)[1:100], all.cpg = rownames(gmset_anno), collection = "KEGG", array.type = "450K")
top100_ACTvREM_kegg <- top100_ACTvREM_kegg[order(top100_ACTvREM_kegg$P.DE),]
write.csv(top100_ACTvREM_kegg, file.path(pathwayDir, "ACTvREM_top100_kegg.csv"))
```

```{r DMR analysis}
dmrDir <- file.path(outputDir, "dmrs")
dir.create(dmrDir)

require(DMRcate)
dmr_anno_CDvNon_CD <- cpg.annotate(object = as.matrix(mvals_filtered),
                                   datatype = "array",
                                   arraytype = "450K",
                                   what = "M",
                                   analysis.type = "differential",
                                   design = design_mat,
                                   contrasts = T,
                                   cont.matrix = contrast_mat,
                                   coef = "CDvNon_CD")
dmr_CDvNon_CD <- dmrcate(dmr_anno_CDvNon_CD, lambda = 1000, C = 2, min.cpgs = 3, mc.cores = 8, pcutoff = 0.05)
dmr_CDvNon_CD_gr <- extractRanges(dmr_CDvNon_CD, genome = "hg19")

write.csv(as.data.frame(dmr_CDvNon_CD_gr), file.path(dmrDir, "dmrs_CDvNon_CD.csv"))

dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity)
```

We observe one DMR that we have observed previously: chr7:51538650-51539678. In this region, CD samples are hypermethylated relative to non-CD samples. Unfortunately, Illumina has no annotation for the probes within this region. One possible reason is that this region falls within an enhancer. We can use the Javierre et al. 2016 data to investigate this possibility. 

```{r Enhancers pcHiC}
pchic <- read.csv("~/archive/common_data/enhancers/javierre2016/ActivePromoterEnhancerLinks_hg19.tsv", sep = "\t")
pchic_gr <- makeGRangesFromDataFrame(pchic, keep.extra.columns = T, seqnames.field = "oeChr", start.field = "oeSt", end.field = "oeEnd")

findOverlaps(pchic_gr, dmr_CDvNon_CD_gr[1])
```
No overlaps were found, but Javierre et al. might not be the only source to look at. We will next Try Ensembl Regulatory. As Ensembl Regulatory uses GRCh38, we will need to liftover the coordinates.

```{r Enhancers Ensembl Regulatory}
require(rtracklayer)
hg19togrch38 <- import.chain("~/archive/common_data/liftover_chains/hg19ToHg38.over.chain")
dmr_CDvNon_CD_gr_grch38 <- liftOver(dmr_CDvNon_CD_gr[1], hg19togrch38)
```

Given that no systematic differences can be observed, we next questioned whether any positions were variably affected (e.g. the variance is increased in one group versus the other)

```{r VMPs}
vmpDir <- file.path(outputDir, "vmps")
dir.create(vmpDir)

require(parallel)
require(car)
require(NDlib)
require(Cairo)

cl <- makeCluster(detectCores()-2)

mult.var.test <- function(i, factor_interest, method = c("levene", "bartlett", "fligner")){
  method <- match.arg(method)
  
  if(method == "levene"){
    pval <- leveneTest(y = i, group = factor_interest)$Pr[1]
    statistic <- leveneTest(y = i, group = factor_interest)$F[1]
  } else if(method == "bartlett"){
    pval <- bartlett.test(x = i, g = factor_interest)$"p.value"
    statistic <- as.numeric(bartlett.test(x = i, g = factor_interest)$"statistic")
  } else if(method == "fligner"){
    pval <- fligner.test(x = i, g = factor_interest)$"p.value"
    statistic <- fligner.test(x = i, g = factor_interest)$"statistic"
  }
  
  results <- data.frame(pval = pval, statistic = statistic)
  return(results)
}

clusterEvalQ(cl, library(car))

var_CDvNon_CD <- parApply(cl = cl, X = mvals_filtered, MARGIN = 1, FUN = mult.var.test, factor_interest = pdata_filtered$Status1)
var_ACTvREM <- parApply(cl = cl, X = mvals_filtered, MARGIN = 1, FUN = mult.var.test, factor_interest = pdata_filtered$Severity)
stopCluster(cl)

#CDvNon_CD
var_CDvNon_CD_df <- do.call(rbind, var_CDvNon_CD)
var_CDvNon_CD_df$padj <- p.adjust(var_CDvNon_CD_df$pval)
var_CDvNon_CD_df <- var_CDvNon_CD_df[order(var_CDvNon_CD_df$pval),]
var_CDvNon_CD_df$Gene <- gmset_anno_gr[rownames(var_CDvNon_CD_df),]$UCSC_RefGene_Name
var_CDvNon_CD_sig <- var_CDvNon_CD_df[var_CDvNon_CD_df$padj < 0.05,]

write.csv(var_CDvNon_CD_df, file.path(vmpDir, "vmp_CDvNon_CD.csv"))

#ACTvREM
var_ACTvREM_df <- do.call(rbind, var_ACTvREM)
var_ACTvREM_df$padj <- p.adjust(var_ACTvREM_df$pval)
var_ACTvREM_df <- var_ACTvREM_df[order(var_ACTvREM_df$pval),]
var_ACTvREM_df$Gene <- gmset_anno_gr[rownames(var_ACTvREM_df),]$UCSC_RefGene_Name
var_ACTvREM_sig <- var_ACTvREM_df[var_ACTvREM_df$padj < 0.05,]

write.csv(var_ACTvREM_df, file.path(vmpDir, "vmp_ACTvREM.csv"))

for(i in 1:10){
  Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(vmpDir, paste0(i, "_", rownames(var_CDvNon_CD_df)[i], "_CDvNon_CD.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(var_CDvNon_CD_df)[i], betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F))
  dev.off()
  
  Cairo(width = 1000, height = 1000, type = "pdf", file = file.path(vmpDir, paste0(i, "_", rownames(var_ACTvREM_df)[i], "_ACTvREM.pdf")), bg = "white", units = "px", dpi = 120)
  print(cpg_strip_plot(cpg_num = rownames(var_ACTvREM_df)[i], betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F))
  dev.off()
}
```

#Figures

```{r Figure 1}
require(ggpubr)
require(Cairo)

fig1Dir <- file.path(manuscriptDir, "fig1")
dir.create(fig1Dir)

fig1A <- cpg_strip_plot(cpg_num = rownames(CDvNon_CD)[1], betas = betas_filtered, factor_interest = pdata_filtered$Status1, enlarged = F, type = "SE") + theme(legend.pos = "none") + labs(subtitle = paste(CDvNon_CD$chr[1], CDvNon_CD$pos[1], sep = ":"))

fig1B <- dmr_plot(dmr_gr = range(cg17206772_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Status1, highlight = CDvNon_CD$pos[1]) + theme(legend.pos = "bottom", legend.title = element_blank())

fig1 <- ggarrange(fig1A, fig1B, ncol = 2, labels = "AUTO", common.legend = T, legend = "bottom", align = "h")

Cairo(width = 1000, height = 750, type = "png", file = file.path(fig1Dir, "fig1.png"), bg = "white", units = "px", dpi = 120)
print(fig1)
dev.off()

Cairo(width = 2000, height = 1500, type = "pdf", file = file.path(fig1Dir, "fig1.pdf"), bg = "white", units = "px", dpi = 120)
print(fig1)
dev.off()
```

```{r Figure 2}
fig2Dir <- file.path(manuscriptDir, "fig2")
dir.create(fig2Dir)

fig2A <- cpg_strip_plot(cpg_num = rownames(CDvNon_CD)[2], betas = betas_filtered, factor_interest = pdata_filtered$Status1, enlarged = F, type = "SE") + theme(legend.pos = "none") + labs(subtitle = paste(CDvNon_CD$chr[2], CDvNon_CD$pos[2], sep = ":"))

fig2B <- dmr_plot(dmr_gr = dmr_CDvNon_CD_gr[1], meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Status1, highlight = CDvNon_CD$pos[2]) + theme(legend.pos = "bottom", legend.title = element_blank())

fig2 <- ggarrange(fig2A, fig2B, ncol = 2, labels = "AUTO", common.legend = T, legend = "bottom", align = "h")

Cairo(width = 1000, height = 750, type = "png", file = file.path(fig2Dir, "fig2.png"), bg = "white", units = "px", dpi = 120)
print(fig2)
dev.off()

Cairo(width = 2000, height = 1500, type = "pdf", file = file.path(fig2Dir, "fig2.pdf"), bg = "white", units = "px", dpi = 120)
print(fig2)
dev.off()
```


```{r Figure 3}
fig3Dir <- file.path(manuscriptDir, "fig3")
dir.create(fig3Dir)

fig3A <- cpg_strip_plot(cpg_num = rownames(ACTvREM)[2], betas = betas_filtered, factor_interest = pdata_filtered$Severity, enlarged = F, type = "SE") + theme(legend.pos = "none") + labs(subtitle = paste(ACTvREM$chr[2], ACTvREM$pos[2], sep = ":"))

fig3B <- dmr_plot(dmr_gr = range(cg06521852_area_gr, ignore.strand = T), meth_data = betas_filtered, anno_gr = gmset_anno_gr, meth_groups = pdata_filtered$Severity, highlight = ACTvREM$pos[2]) + theme(legend.pos = "bottom", legend.title = element_blank())

fig3 <- ggarrange(fig3A, fig3B, ncol = 2, labels = "AUTO", common.legend = T, legend = "bottom", align = "h")

Cairo(width = 1000, height = 750, type = "png", file = file.path(fig3Dir, "fig3.png"), bg = "white", units = "px", dpi = 120)
print(fig3)
dev.off()

Cairo(width = 2000, height = 1500, type = "pdf", file = file.path(fig3Dir, "fig3.pdf"), bg = "white", units = "px", dpi = 120)
print(fig3)
dev.off()
```

